[![Support Site Activity](https://bioconductor.org/shields/posts/curatedMetagenomicData.svg)](https://support.bioconductor.org/t/curatedmetagenomicdata/)
[![Build Results](https://bioconductor.org/shields/build/devel/data-experiment/curatedMetagenomicData.svg)](https://bioconductor.org/checkResults/devel/data-experiment-LATEST/curatedMetagenomicData/)
[![Travis-CI Build Status](https://travis-ci.org/waldronlab/curatedMetagenomicData.svg?branch=master)](https://travis-ci.org/waldronlab/curatedMetagenomicData)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/waldronlab/curatedMetagenomicData?branch=master&svg=true)](https://ci.appveyor.com/project/schifferl/curatedmetagenomicdata-o9eib)
[![Coverage Status](https://img.shields.io/codecov/c/github/waldronlab/curatedMetagenomicData/master.svg)](https://codecov.io/github/waldronlab/curatedMetagenomicData?branch=master)
[![Subversion Commits](https://bioconductor.org/shields/commits/data-experiment/curatedMetagenomicData.svg)](https://bioconductor.org/packages/devel/data/experiment/html/curatedMetagenomicData.html#svn_source)

# curatedMetagenomicData

*curatedMetagenomicData* is a *Bioconductor* package that uses *ExperimentHub* to access
a large number (n â‰ˆ 3000) of human microbiome samples gathered from various sources.

## Quick Start

Install `curatedMetagenomicData` through Bioconductor using `BiocInstaller` and open the [package vignette](https://bioconductor.org/packages/release/data/experiment/vignettes/curatedMetagenomicData/inst/doc/curatedMetagenomicData.html).

```
BiocInstaller::biocLite("curatedMetagenomicData")
browseVignettes("curatedMetagenomicData")
```

For more extensive information on installing and using `curatedMetagenomicData` please see https://waldronlab.github.io/curatedMetagenomicData/.

## Adding New Data

To add new data there a number of developer functions, the details of which are explained in the [wiki](https://github.com/waldronlab/curatedMetagenomicData/wiki/Adding-New-Data).
