---
title: "Taxanomic and Pathway Abundance in All Gut Samples"
author: "Audrey Renson"
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{Taxanomic and Pathway Abundance in All Gut Samples}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

```{r libraries, include=FALSE}
library(curatedMetagenomicData)
library(RColorBrewer)
library(ggplot2)
library(stats)
library(reshape2)
```

# Coordinated Color Scheme

```{r}
blue <- "#3366aa"
blueGreen <- "#11aa99"
green <- "#66aa55"
paleYellow <- "#cccc55"
gray <- "#777777"
purple <- "#992288"
red <- "#ee3333"
orange <- "#ee7722"
yellow <- "#ffee33"
pallet <- c(blue, blueGreen, green, paleYellow, gray, purple, red, orange, yellow)
n <- length(pallet)
image(1:n, 1, as.matrix(1:n), col = pallet, xlab = "", ylab = "", xaxt = "n", yaxt = "n", bty = "n")
```

# Assemble pathway abundance dataset

## Retrieve data

```{r}
curatedMetagenomicData <- function(x = "*", dryrun = TRUE, counts = FALSE, bugs.as.phyloseq = FALSE, x.is.glob = TRUE) {
    requested.datasets <- x
    all.datasets <- ls("package:curatedMetagenomicData")
    all.datasets <- grep("marker|gene|path|metaphlan_bugs", all.datasets, val = TRUE)
    regex <- ifelse(x.is.glob, paste(glob2rx(requested.datasets), collapse = "|"), requested.datasets)
    matched.datasets <- grep(regex, all.datasets, value = TRUE)
    if (dryrun) {
        message(c(
            "Dry run: see return values for datasets that would be downloaded.",
            "\nRun with `dryrun=FALSE` to actually download these datasets."))
        return(matched.datasets)
    }
    if (!any(matched.datasets %in% all.datasets)) 
        stop("requested datasets do not match any available datasets.")
    eset.list <- lapply(seq_along(matched.datasets), function(i) {
        message(paste0("Working on ", matched.datasets[i]))
        eset <- do.call(get(matched.datasets[i]), list())
        if (counts) {
            exprs(eset) <- round(sweep(exprs(eset), 2, eset$number_reads/100, "*"))
        }
        if (bugs.as.phyloseq && grepl("metaphlan", matched.datasets[i])) {
            eset <- ExpressionSet2phyloseq(eset)
        }
        return(eset)
    })
    names(eset.list) <- matched.datasets
    return(eset.list)
}
```

Create a list of ExpressionSet objects:

```{r}
eset.list <- curatedMetagenomicData("*pathabundance_relab.stool", dryrun = FALSE)
```

Give them simplified titles:

```{r}
names(eset.list) <- gsub("\\..+", "", names(eset.list))
```

And add the titles to the colnames:

```{r}
for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <- paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <- pData(eset.list[[i]])[, !sapply(pData(eset.list[[i]]), function(x) all(is.na(x)))]
    eset.list[[i]]$subjectID <- as.character(eset.list[[i]]$subjectID)
}
```

Remove rows providing genus/species names:

```{r}
for (i in seq_along(eset.list)) {
    eset.list[[i]] <- eset.list[[i]][!grepl("\\|", rownames(eset.list[[i]])), ]
}
```

## Merge into one big `ExpressionSet`

```{r}
joinWithRnames <- function(obj, FUN = I) {
    mylist <- lapply(obj, function(x) {
        df <- data.frame(FUN(x))
        df$rnames28591436107 <- rownames(df)
        return(df)
    })
    bigdf <- Reduce(full_join, mylist)
    rownames(bigdf) <- make.names(bigdf$rnames28591436107)
    bigdf <- bigdf[, !grepl("^rnames28591436107$", colnames(bigdf))]
    return(bigdf)
}

pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0
eset_pathway <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))
```

# Assemble bug abundance dataset

## Retrieve data

Create a list of ExpressionSet objects:

```{r}
eset.list <- curatedMetagenomicData("*metaphlan_bugs_list.stool", dryrun = FALSE)
```

Give them simplified titles:

```{r}
names(eset.list) <- gsub("\\..+", "", names(eset.list))
```

And add the titles to the colnames:

```{r}
for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <- paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <- pData(eset.list[[i]])[, !sapply(pData(eset.list[[i]]), function(x) all(is.na(x)))]
    eset.list[[i]]$subjectID <- as.character(eset.list[[i]]$subjectID)
}
```

Keep leaf nodes only:

```{r}
for (i in seq_along(eset.list)) {
    eset.list[[i]] <- eset.list[[i]][grep("t__", rownames(eset.list[[i]]), invert = TRUE), ]
    eset.list[[i]] <- eset.list[[i]][grep("s__|_unclassified\t", rownames(eset.list[[i]]), perl = TRUE), ]
}
```

## Merge into one big `ExpressionSet`

```{r}
pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0
eset_bugs <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))
```

## Convert bug abundance `ExpressionSet` to `phyloseq` object

```{r}
metaphlanToPhyloseq <- function(tax, metadat = NULL, simplenames = TRUE, roundtointeger = FALSE, split = "|") {
    xnames <- rownames(tax)
    shortnames <- gsub(paste0(".+\\", split), "", xnames)
    if (simplenames) {
        rownames(tax) <- shortnames
    }
    if (roundtointeger) {
        tax <- round(tax * 10000)
    }
    x2 <- strsplit(xnames, split = split, fixed = TRUE)
    taxmat <- matrix(NA, ncol = max(sapply(x2, length)), nrow = length(x2))
    colnames(taxmat) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "Strain")[1:ncol(taxmat)]
    rownames(taxmat) <- rownames(tax)
    for (i in 1:nrow(taxmat)) {
        taxmat[i, 1:length(x2[[i]])] <- x2[[i]]
    }
    taxmat <- gsub("[a-z]__", "", taxmat)
    taxmat <- phyloseq::tax_table(taxmat)
    otutab <- phyloseq::otu_table(tax, taxa_are_rows = TRUE)
    if (is.null(metadat)) {
        res <- phyloseq::phyloseq(taxmat, otutab)
    } else {
        res <- phyloseq::phyloseq(taxmat, otutab, phyloseq::sample_data(metadat))
    }
    return(res)
}

pseq <- metaphlanToPhyloseq(tax = exprs(eset_bugs), metadat = pData(eset_bugs), split = ".")
```

# Within-sample relative bug abundances
```{r}
glom <- tax_glom(pseq, taxrank = "Phylum")

top8phyla = names(sort(taxa_sums(glom), TRUE)[1:8])
phyla8_subset <- prune_taxa(top8phyla, glom)

phyla_to_sort <- data.frame(id = 1:8, phyla = as.character(tax_table(phyla8_subset)[, "Phylum"]), otu = as.character(taxa_names(phyla8_subset)))
rownames(phyla_to_sort) <- phyla_to_sort$otu

phylum_ranks <- phyla8_subset %>%
    otu_table %>%
    rowSums %>%
    sort(T) %>%
    names

phyla_to_sort <- phyla_to_sort[phylum_ranks, ]

prop <- transform_sample_counts(phyla8_subset, function(i) i/sum(i))

bardat <- psmelt(prop) %>%
    select(OTU, Sample, Abundance) %>%
    mutate(Sample = as.numeric(factor(Sample)), OTU = factor(OTU, levels = phyla_to_sort$otu, labels = phyla_to_sort$phyla))

firmicutes_order <- bardat %>%
    filter(OTU == "Firmicutes") %>%
    arrange(Abundance) %>%
    select(Sample)

bardat %<>%
    mutate(Sample = as.numeric(factor(Sample, levels = factor(firmicutes_order$Sample))))

set.seed(14)

bardat %>%
    arrange(desc(OTU), Abundance) %>%
    ggplot(aes(x = Sample, y = Abundance, fill = OTU)) +
    geom_area() +
    scale_fill_manual(values = sample(pallet, size = 8, replace = FALSE)) +
    labs(fill = "Phylum") +
    theme(axis.text.x = element_blank(), legend.position = "bottom")
```

# Alpha diversity by study

```{r}
alpha <- estimate_richness(pseq, measures = "Shannon")
alpha$study <- sample_data(pseq)$study

# sort by group median

alpha %<>%
    group_by(study) %>%
    mutate(median = median(Shannon)) %>%
    arrange(desc(median)) %>%
    ungroup %>%
    mutate(study_num = as.numeric(as.factor(alpha$study)))

box_order <- factor(unique(alpha$study[order(alpha$median)]))

alpha$study <- factor(alpha$study, levels = box_order)

alpha %>%
    ggplot(aes(x = study, y = Shannon, fill = study)) +
    stat_boxplot(geom = "errorbar") +
    geom_boxplot() +
    theme(axis.text.x = element_blank()) +
    ylab("Shannon Alpha Diversity") +
    xlab("")
```

# Correlation of metabolic pathways with most abundant bugs

## With Prevotella *copri*

```{r}
eset_pathway$prev <- as.numeric(exprs(eset_bugs)[grep("s__Prevotella_copri", rownames(exprs(eset_bugs))), ])

cor_est_p <- function(x1, x2) {
    cor <- cor.test(x1, x2)
    c(r = cor$estimate, p = cor$p.value)
}

cors <- t(sapply(featureNames(eset_pathway), function(i) cor_est_p(exprs(eset_pathway)[i, ], eset_pathway$prev)))

feature <- rownames(cors)

cors <- as.data.frame(cors)
cors$feature <- feature
cors <- na.omit(cors)

par(mar = c(10, 10, 10, 10))
qplot(x = eset_pathway$prev,
      y = exprs(eset_pathway)[cors$feature[cors$r.cor == max(cors$r.cor)], ],
      xlab = "Prevotella copri",
      ylab = "Pathway abundance",
      main = "PWY 7291: Adenosine ribonucleotides \nde novo biosynthesis ",
      colour = I("dodgerblue"),
      shape = I(1))
```

## With top 20 most abundant genera

```{r, fig.width=10}
glom_genus <- tax_glom(pseq, taxrank = "Genus")
top20_genus_otu_names <- names(sort(taxa_sums(glom_genus), TRUE)[1:20])
top20_genus <- tax_table(glom_genus)[top20_genus_otu_names, "Genus"]
subset_genus <- prune_taxa(top20_genus_otu_names, glom_genus)

max_cor_pathway <- function(y, X, margin = 1) {
    cors <- apply(X, margin, function(x) cor.test(as.numeric(y), as.numeric(x))$estimate)
    rownames(X)[cors == max(cors)]
}

max_pathways <- apply(otu_table(subset_genus), 1, function(y) max_cor_pathway(y, X = exprs(eset_pathway)[1:20, ]))

cor_matrix <- cor(t(otu_table(subset_genus)), t(exprs(eset_pathway)[max_pathways, ]))
rownames(cor_matrix) <- top20_genus[rownames(cor_matrix), 1]

melted_cors <- melt(cor_matrix)

melted_cors %>%
    ggplot(aes(x = Var1, y = Var2, fill = value)) +
    geom_tile() +
    scale_fill_gradient2(low = blueGreen, high = purple, mid = "gray90", midpoint = 0, space = "Lab", name = "Pearson\nCorrelation") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 9), axis.text.y = element_text(size = 9)) +
    labs(x = "Genus", y = "Pathway")
```
