---
title: "Taxanomic and Pathway Abundance in All Gut Samples"
author: "Audrey Renson"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    number_sections: yes
    toc: yes
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Fetch and Cluster Gut Metagenomic Taxonomic Tables}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE)
```
# Assemble pathway abundance dataset
## Retrieve data
Create an ExperimentHub containing metadata for all `curatedMetagenomicData` records:

```{r}
suppressPackageStartupMessages(library(ExperimentHub))
eh = ExperimentHub()
myquery = query(eh, "curatedMetagenomicData")
```

Subset this to bug abundance from stool datasets. Note that this could also be done using `myquery$tags` (once the tags are updated). See the available metadata using `View(mcols(myquery))`.

```{r}
myquery.stool <- myquery[grepl("stool", myquery$title) & grepl("pathabundance", myquery$title), ]
myquery.stool
```

Create a list of ExpressionSet objects:
```{r, message=FALSE}
eset.list <- lapply(names(myquery.stool), function(x) myquery.stool[[x]])
```

Give them simplified titles:
```{r}
names(eset.list) <- myquery.stool$title
names(eset.list) <- gsub("\\..+", "", myquery.stool$title)
```

And add the titles to the colnames:
```{r}
for (i in 1:length(eset.list)){
    colnames(eset.list[[i]]) <- paste(names(eset.list)[[i]],
                                      colnames(eset.list[[i]]), 
                                      sep=".")
    pData(eset.list[[i]]) <- pData(eset.list[[i]])[, !sapply(pData(eset.list[[i]]), function(x) all(is.na(x)))]
    eset.list[[i]]$subjectID <- as.character(eset.list[[i]]$subjectID)
}
```
Keep leaf nodes only:
```{r}
for (i in seq_along(eset.list)){
  eset.list[[i]] <- eset.list[[i]][grep("t__", rownames(eset.list[[i]]), invert=TRUE), ]
  eset.list[[i]] <- eset.list[[i]][grep("s__|_unclassified\t", rownames(eset.list[[i]]), perl=TRUE), ]
}
```

## Merge into one big `ExpressionSet`

```{r join, message=FALSE, cache=TRUE}
joinWithRnames <- function(obj, FUN = I) {
    mylist <- lapply(obj, function(x) {
        df <- data.frame(FUN(x))
        df$rnames28591436107 <- rownames(df)
        return(df)
    })
    bigdf <- Reduce(dplyr::full_join, mylist)
    rownames(bigdf) <- make.names(bigdf$rnames28591436107)
    bigdf <- bigdf[, !grepl("^rnames28591436107$", colnames(bigdf))]
    return(bigdf)
}

pdat <- joinWithRnames(eset.list, FUN=pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN=exprs)
ab[is.na(ab)] <- 0
eset_pathway <- ExpressionSet(assayData = as.matrix(ab), 
                      phenoData = AnnotatedDataFrame(pdat))
eset_pathway
```

# Assemble bug abundance dataset

## Retrieve data

```{r}
myquery.stool <- myquery[grepl("stool", myquery$title) & grepl("bugs", myquery$title), ]
myquery.stool
```

Create a list of ExpressionSet objects:
```{r, message=FALSE}
eset.list <- lapply(names(myquery.stool), function(x) myquery.stool[[x]])
```

Give them simplified titles:
```{r}
names(eset.list) <- myquery.stool$title
names(eset.list) <- gsub("\\..+", "", myquery.stool$title)
```

And add the titles to the colnames:
```{r}
for (i in 1:length(eset.list)){
    colnames(eset.list[[i]]) <- paste(names(eset.list)[[i]],
                                      colnames(eset.list[[i]]), 
                                      sep=".")
    pData(eset.list[[i]]) <- pData(eset.list[[i]])[, !sapply(pData(eset.list[[i]]), function(x) all(is.na(x)))]
    eset.list[[i]]$subjectID <- as.character(eset.list[[i]]$subjectID)
}
```
Keep leaf nodes only:
```{r}
for (i in seq_along(eset.list)){
  eset.list[[i]] <- eset.list[[i]][grep("t__", rownames(eset.list[[i]]), invert=TRUE), ]
  eset.list[[i]] <- eset.list[[i]][grep("s__|_unclassified\t", rownames(eset.list[[i]]), perl=TRUE), ]
}
```

## Merge into one big `ExpressionSet`

```{r join_bugs, message=FALSE, cache=TRUE}
joinWithRnames <- function(obj, FUN = I) {
    mylist <- lapply(obj, function(x) {
        df <- data.frame(FUN(x))
        df$rnames28591436107 <- rownames(df)
        return(df)
    })
    bigdf <- Reduce(dplyr::full_join, mylist)
    rownames(bigdf) <- make.names(bigdf$rnames28591436107)
    bigdf <- bigdf[, !grepl("^rnames28591436107$", colnames(bigdf))]
    return(bigdf)
}

pdat <- joinWithRnames(eset.list, FUN=pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN=exprs)
ab[is.na(ab)] <- 0
eset_bugs <- ExpressionSet(assayData = as.matrix(ab), 
                      phenoData = AnnotatedDataFrame(pdat))
eset_bugs
```
## Convert bug abundance `ExpressionSet` to `phyloseq` object

```{r}
suppressPackageStartupMessages(library(phyloseq))
source("https://raw.githubusercontent.com/waldronlab/presentations/master/Waldron_2016-06-07_EPIC/metaphlanToPhyloseq.R")
pseq <- metaphlanToPhyloseq(tax=exprs(eset_bugs), metadat=pData(eset_bugs), split=".")
```


```{r}
##Coordinated color scheme
source("https://raw.githubusercontent.com/waldronlab/curatedMetagenomicData/master/vignettes/extras/colorPalette_curatedMetagenomicData.R")

```

# Within-sample relative bug abundances

```{r barplot_relative_abundance}
suppressPackageStartupMessages({
 library(RColorBrewer)
 library(dplyr)
 library(ggplot2)
})

#glom by phylum
glom <- tax_glom(pseq, taxrank = "Phylum")

#subset most abundant 8 phyla
top8phyla = names(sort(taxa_sums(glom), TRUE)[1:8])
phyla8_subset <- prune_taxa(top8phyla, glom)

#sort by phylum abundance
phyla_to_sort <- data.frame(id=1:8, phyla = as.character(tax_table(phyla8_subset)[,"Phylum"]), 
                            otu = as.character(taxa_names(phyla8_subset)))
rownames(phyla_to_sort) <- phyla_to_sort$otu
phylum_ranks <- phyla8_subset %>% otu_table %>% rowSums %>% sort(T) %>% names
phyla_to_sort <- phyla_to_sort[phylum_ranks, ]

#calculate abundance as a proportion of top 8
prop <- transform_sample_counts(phyla8_subset, function(i) i / sum(i))

#melt for use in plotting
bardat <- psmelt(prop) %>% select(OTU, Sample, Abundance) %>%  
  mutate(Sample = as.numeric(factor(Sample)),    # geom_area only seems to work with a numeric x
         OTU = factor(OTU,                         # sort by phyla
                      levels=phyla_to_sort$otu,   
                      labels=phyla_to_sort$phyla)) 


firmicutes_order <- bardat %>%     # get order by firmicutes abundance
  filter(OTU=="Firmicutes") %>% 
  arrange(Abundance) %>% 
  select(Sample)

bardat %<>% mutate(Sample = as.numeric(factor(Sample, levels=factor(firmicutes_order$Sample)))) # apply firmicutes order to data.frame
  

#barplot
set.seed(14)
b <- ggplot(arrange(bardat, desc(OTU), Abundance), aes(x=Sample, y=Abundance, fill=OTU))
b <- b + geom_area()
b <- b + scale_fill_manual(values=sample(pallet, size = 8, replace=F))
b <- b + labs(fill="Phylum")
b + theme(axis.text.x = element_blank(), legend.position="bottom")

```


# Alpha diversity by study
```{r alpha_estimate}

alpha <- estimate_richness(pseq, measures = "Shannon")
alpha$study = sample_data(pseq)$study

#sort by group median

alpha %<>%
  group_by(study) %>% mutate(median = median(Shannon)) %>% arrange(desc(median)) %>% 
  ungroup %>%  mutate(study_num = as.numeric(as.factor(alpha$study)))

box_order <- factor(unique(alpha$study[order(alpha$median)]))

alpha$study <- factor(alpha$study, levels=box_order)

p <- ggplot(alpha, aes(x=study, y=Shannon, fill=study))
p <- p + stat_boxplot(geom ='errorbar') 
p <- p + geom_boxplot() 
p <- p + theme(axis.text.x = element_blank())
p <- p + ylab("Shannon Alpha Diversity") + xlab("") 
p
```

# Correlation of metabolic pathways with most abundant bugs

## With Prevotella *copri*
```{r warning=F}
eset_pathway$prev <- as.numeric(exprs(eset_bugs)[grep("s__Prevotella_copri",rownames(exprs(eset_bugs))),] )

cor_est_p <- function(x1,x2) {
  cor <- cor.test(x1,x2)
  c(r=cor$estimate, p=cor$p.value)
}
 
cors <- t(sapply(featureNames(eset_pathway), function(i) cor_est_p(exprs(eset_pathway)[i,], eset_pathway$prev)))

feature <- rownames(cors) 

cors <- as.data.frame(cors)
cors$feature <- feature
cors <- na.omit(cors)

suppressPackageStartupMessages(library(ggplot2))


par(mar=c(10,10,10,10))
qplot(x    = eset_pathway$prev, 
     y    = exprs(eset_pathway)[cors$feature[cors$r.cor == max(cors$r.cor)]  , ], 
     xlab = "Prevotella copri",
     ylab = "Pathway abundance",
     main = "PWY 7291: Adenosine ribonucleotides \nde novo biosynthesis ",
     colour = I("dodgerblue"), shape=I(1))


```

## With top 20 most abundant genera
```{r heatmap_cors, fig.width=10}

#get top 20 genuses
glom_genus <- tax_glom(pseq, taxrank = "Genus")
top20_genus_otu_names = names(sort(taxa_sums(glom_genus), TRUE)[1:20])
top20_genus <- tax_table(glom_genus)[top20_genus_otu_names,"Genus"]
subset_genus <- prune_taxa(top20_genus_otu_names, glom_genus)

max_cor_pathway <- function(y, X, margin=1) {
  #X is a matrix
  cors = apply(X, margin, function(x) cor.test(as.numeric(y), as.numeric(x))$estimate)
  rownames(X)[cors==max(cors)]
}

max_pathways <- apply(otu_table(subset_genus), 1,  function(y) max_cor_pathway(y, X=exprs(eset_pathway)[1:20,]))

cor_matrix <- stats::cor(t(otu_table(subset_genus)), t(exprs(eset_pathway)[max_pathways,]))
rownames(cor_matrix) = top20_genus[rownames(cor_matrix), 1]

melted_cors <- reshape2::melt(cor_matrix)
```

```{r heatmap_plot, fig.width=10}
ggplot(data = melted_cors, aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile() +
  scale_fill_gradient2(low = blueGreen, high = purple, mid = 'gray90', 
   midpoint = 0, space = "Lab", 
   name="Pearson\nCorrelation") + 
  theme(axis.text.x = element_text(angle = 45, vjust=1, hjust = 1, size=9), axis.text.y=element_text(size=9)) +
  labs(x="Genus",y="Pathway")
```



