---
title: "Clustering of All Gut Samples by Taxonomic Abundance"
author: "Audrey Renson"
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{Clustering of All Gut Samples by Taxonomic Abundance}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

```{r libraries, include=FALSE}
library(curatedMetagenomicData)
library(ggplot2)
library(cluster)
library(fpc)
```

# Coordinated Color Scheme

```{r}
blue <- "#3366aa"
blueGreen <- "#11aa99"
green <- "#66aa55"
paleYellow <- "#cccc55"
gray <- "#777777"
purple <- "#992288"
red <- "#ee3333"
orange <- "#ee7722"
yellow <- "#ffee33"
pallet <- c(blue, blueGreen, green, paleYellow, gray, purple, red, orange, yellow)
n <- length(pallet)
image(1:n, 1, as.matrix(1:n), col = pallet, xlab = "", ylab = "", xaxt = "n", yaxt = "n", bty = "n")
```

# Create datasets

```{r}
curatedMetagenomicData <- function(x = "*", dryrun = TRUE, counts = FALSE, bugs.as.phyloseq = FALSE, x.is.glob = TRUE) {
    requested.datasets <- x
    all.datasets <- ls("package:curatedMetagenomicData")
    all.datasets <- grep("marker|gene|path|metaphlan_bugs", all.datasets, val = TRUE)
    regex <- ifelse(x.is.glob, paste(glob2rx(requested.datasets), collapse = "|"), requested.datasets)
    matched.datasets <- grep(regex, all.datasets, value = TRUE)
    if (dryrun) {
        message(c(
            "Dry run: see return values for datasets that would be downloaded.",
            "\nRun with `dryrun=FALSE` to actually download these datasets."))
        return(matched.datasets)
    }
    if (!any(matched.datasets %in% all.datasets)) 
        stop("requested datasets do not match any available datasets.")
    eset.list <- lapply(seq_along(matched.datasets), function(i) {
        message(paste0("Working on ", matched.datasets[i]))
        eset <- do.call(get(matched.datasets[i]), list())
        if (counts) {
            exprs(eset) <- round(sweep(exprs(eset), 2, eset$number_reads/100, "*"))
        }
        if (bugs.as.phyloseq && grepl("metaphlan", matched.datasets[i])) {
            eset <- ExpressionSet2phyloseq(eset)
        }
        return(eset)
    })
    names(eset.list) <- matched.datasets
    return(eset.list)
}
```

Create a list of ExpressionSet objects:

```{r}
eset.list <- curatedMetagenomicData("*metaphlan_bugs_list.stool", dryrun = FALSE)
```

Give them simplified titles:

```{r}
names(eset.list) <- gsub("\\..+", "", names(eset.list))
```

And add the titles to the colnames:

```{r}
for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <- paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <- pData(eset.list[[i]])[, !sapply(pData(eset.list[[i]]), function(x) all(is.na(x)))]
    eset.list[[i]]$subjectID <- as.character(eset.list[[i]]$subjectID)
}
```

Keep leaf nodes only:

```{r}
for (i in seq_along(eset.list)) {
    eset.list[[i]] <- eset.list[[i]][grep("t__", rownames(eset.list[[i]]), invert = TRUE), ]
    eset.list[[i]] <- eset.list[[i]][grep("s__|_unclassified\t", rownames(eset.list[[i]]), perl = TRUE), ]
}
```

# Merging into one big `ExpressionSet`

```{r}
joinWithRnames <- function(obj, FUN = I) {
    mylist <- lapply(obj, function(x) {
        df <- data.frame(FUN(x))
        df$rnames28591436107 <- rownames(df)
        return(df)
    })
    bigdf <- Reduce(full_join, mylist)
    rownames(bigdf) <- make.names(bigdf$rnames28591436107)
    bigdf <- bigdf[, !grepl("^rnames28591436107$", colnames(bigdf))]
    return(bigdf)
}

pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))

ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0

eset <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))
```

# Convert to `phyloseq` object

```{r}
metaphlanToPhyloseq <- function(tax, metadat = NULL, simplenames = TRUE, roundtointeger = FALSE, split = "|") {
    xnames <- rownames(tax)
    shortnames <- gsub(paste0(".+\\", split), "", xnames)
    if (simplenames) {
        rownames(tax) <- shortnames
    }
    if (roundtointeger) {
        tax <- round(tax * 10000)
    }
    x2 <- strsplit(xnames, split = split, fixed = TRUE)
    taxmat <- matrix(NA, ncol = max(sapply(x2, length)), nrow = length(x2))
    colnames(taxmat) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "Strain")[1:ncol(taxmat)]
    rownames(taxmat) <- rownames(tax)
    for (i in 1:nrow(taxmat)) {
        taxmat[i, 1:length(x2[[i]])] <- x2[[i]]
    }
    taxmat <- gsub("[a-z]__", "", taxmat)
    taxmat <- phyloseq::tax_table(taxmat)
    otutab <- phyloseq::otu_table(tax, taxa_are_rows = TRUE)
    if (is.null(metadat)) {
        res <- phyloseq::phyloseq(taxmat, otutab)
    } else {
        res <- phyloseq::phyloseq(taxmat, otutab, phyloseq::sample_data(metadat))
    }
    return(res)
}

pseq <- metaphlanToPhyloseq(tax = exprs(eset), metadat = pData(eset), split = ".")
```

# Principle Coordinates Analysis (PCoA) Plots

## Colored by whether the data are from community samples or the Human Microbiome Project.

```{r create_HMP_vs_community_covariate}
samp <- data.frame(sample_data(pseq))
samp$source <- factor(samp$study == "HMP_2012", levels = c(T, F), labels = c("HMP", "Community"))
sample_data(pseq) <- samp
```

```{r dissimilarity_matrices}
dist_bray <- distance(pseq, method = "bray")
dist_js <- distance(pseq, method = "jsd")
dist_rjs <- sqrt(dist_js)
```

```{r PCoA_ordination}
ord_bray <- ordinate(pseq, method = "PCoA", distance = dist_bray)
ord_JS <- ordinate(pseq, method = "PCoA", distance = dist_js)
ord_RJS <- ordinate(pseq, method = "PCoA", distance = dist_rjs)
```

```{r plot_PCoA}
plot_ordination(pseq, ord_bray, color = "source") +
    ggtitle("Bray-Curtis Principal Coordinates Analysis")

plot_ordination(pseq, ord_JS, color = "source") +
    ggtitle("Jensen-Shannon Principal Coordinates Analysis")

plot_ordination(pseq, ord_RJS, color = "source") +
    ggtitle("Root Jensen-Shannon Principal Coordinates Analysis")
```

## Colored by cluster

```{r PCoA_bray234}
samp$bray_cluster_2 <- factor(pam(dist_bray, k = 2, cluster.only = T))
samp$bray_cluster_3 <- factor(pam(dist_bray, k = 3, cluster.only = T))
samp$bray_cluster_4 <- factor(pam(dist_bray, k = 4, cluster.only = T))
sample_data(pseq) <- samp

plot_ordination(pseq, ord_bray, color = "bray_cluster_2") +
    ggtitle("Bray-Curtis PCoA with 2 Clusters")

plot_ordination(pseq, ord_bray, color = "bray_cluster_3") +
    ggtitle("Bray-Curtis PCoA with 3 Clusters")

plot_ordination(pseq, ord_bray, color = "bray_cluster_4") +
    ggtitle("Bray-Curtis PCoA with 4 Clusters")
```

## Colored by Prevotella copri

```{r PCoA_Prevotella}
Prev <- as.numeric(otu_table(pseq)["s__Prevotella_copri", ])
samp$Prevotella_copri <- Prev
sample_data(pseq) <- samp
plot_ordination(pseq, ord_bray, color = "Prevotella_copri")
```

# Cluster Validation

## Prediction Strength

```{r calculate_Prediction_strength}
ps_bray <- prediction.strength(dist_bray, Gmin = 2, Gmax = 10, clustermethod = pamkCBI)
ps_js <- prediction.strength(dist_js, Gmin = 2, Gmax = 10, clustermethod = pamkCBI)
ps_rjs <- prediction.strength(dist_rjs, Gmin = 2, Gmax = 10, clustermethod = pamkCBI)
```

```{r plot_function}
plot_cluster_validation = function(bray, js, rjs, legend = T, ...) {
    plot(2:10, bray, type = "b", pch = 1, xlab = "Number of Clusters", ...)
    lines(2:10, js, type = "b", pch = 2, lty = 2)
    lines(2:10, rjs, type = "b", pch = 22, lty = 3)
    if (legend) {
        legend("topright", legend = c("Bray-Curtis", "Jensen-Shannon", "Root Jensen-Shannon"), pch = c(1, 2, 22), lty = 1:3)
    }
}
```

```{r plot_Prediction_Strength}
plot_cluster_validation(ps_bray$mean.pred[2:10], ps_js$mean.pred[2:10], ps_rjs$mean.pred[2:10], ylab = "Prediction Strength", ylim = c(0, 1.1), legend = F)
abline(0.9, 0, lty = 5, col = "grey70")
abline(0.8, 0, lty = 8, col = "grey70")
text("Strong support", x = 9, y = 1, col = "grey70")
text("Moderate support", x = 9, y = 0.85, col = "grey70")
text("Little or no support", x = 9, y = 0.6, col = "grey70")
```

## Silhouette Index

```{r Silhouette_Index}
pam_bray <- sapply(2:10, function(i) pam(dist_bray, k = i, cluster.only = T))
pam_js <- sapply(2:10, function(i) pam(dist_js, k = i, cluster.only = T))
pam_rjs <- sapply(2:10, function(i) pam(dist_rjs, k = i, cluster.only = T))

si_bray <- apply(pam_bray, 2, function(i) mean(silhouette(i, dist_bray)[, 3]))
si_js <- apply(pam_js, 2, function(i) mean(silhouette(i, dist_js)[, 3]))
si_rjs <- apply(pam_rjs, 2, function(i) mean(silhouette(i, dist_rjs)[, 3]))
```

```{r plot_Silhouette}
plot_cluster_validation(si_bray, si_js, si_rjs, legend = F, ylab = "Average silhouette width", ylim = c(0, 1))
abline(0.75, 0, lty = 5, col = "grey70")
abline(0.5, 0, lty = 5, col = "grey70")
abline(0.25, 0, lty = 5, col = "grey70")
text("Strong support", x = 9, y = 0.8, col = "grey70")
text("Moderate support", x = 9, y = 0.6, col = "grey70")
text("Weak and could be artificial", x = 8.5, y = 0.4, col = "grey70")
text("No substantial structure", x = 8.5, y = 0.2, col = "grey70")
```

## Calinski-Harabasz

```{r Calinski-Harabasz}
ch_bray <- apply(pam_bray, 2, function(i) cluster.stats(dist_bray, i)$ch)
ch_js <- apply(pam_js, 2, function(i) cluster.stats(dist_js, i)$ch)
ch_rjs <- apply(pam_rjs, 2, function(i) cluster.stats(dist_rjs, i)$ch)
```

```{r plot_Calinski}
plot_cluster_validation(ch_bray, ch_js, ch_rjs, legend = T, ylab = "Calinski-Harabasz score", ylim = c(0, 300))
```

## PCoA k = 2 & gradient for Prevotella & Bacteroides

```{r PCoA_prevotella_bacteroides}
pc1 <- ord_bray$vectors[, 1]
pc2 <- ord_bray$vectors[, 2]

otu_tax <- attr(otu_table(pseq), "dimnames")[[1]]
otu_bacteroides <- otu_table(pseq)[grep("s__Bacteroides", otu_tax), ]
sum_bacteroides <- apply(otu_bacteroides, 2, sum)

df_ordinate <- data.frame(pc1, pc2, bact = sum_bacteroides, prev = Prev, bray2 = as.numeric(samp$bray_cluster_2) + 20)
df_bact <- df_ordinate[df_ordinate$bray2 == 21, ]
df_prev <- df_ordinate[df_ordinate$bray2 == 22, ]

ggplot() +
    geom_point(data = df_prev, aes(x = pc1, y = pc2, shape = factor(bray2), fill = prev), shape = 21, size = 4) +
    scale_fill_gradient(low = "white", high = purple, guide = guide_colorbar(title = "Prevotella copri \n(cluster 2)")) +
    geom_point(data = df_bact, aes(x = pc1, y = pc2, shape = factor(bray2), color = bact), shape = 22, size = 4) +
    scale_color_gradient(low = gray, high = blueGreen, guide = guide_colorbar(title = "Bacteroides \n(cluster 1)")) +
    labs(x = "Axis 1", y = "Axis 2", title = "PCoA by selected abundances, displaying 2 clusters")
```

```{r PCoA_dataset_diseasestate, fig.width=11}
samp$disease[samp$disease %in% c("obesity", "obese")] <- "obesity"
samp$disease[samp$disease %in% c("underweight", "leaness")] <- "underweight"
samp$disease_simplified[samp$disease == "cancer"] <- "cancer"
samp$disease_simplified[samp$disease %in% c("small_adenoma", "large_adenoma")] <- "adenoma"
samp$disease_simplified[samp$disease == "cirrhosis"] <- "cirrhosis"
samp$disease_simplified[samp$disease %in% c("t2d", "impaired_glucose_tolerance")] <- "t2d / impaired glucose tolerance"
samp$disease_simplified[samp$disease %in% c("ibd_crohn_disease", "ibd_ulcerative_colitis")] <- "ibd"
samp$disease_simplified[samp$disease %in% c("obesity", "overweight")] <- "obese or overweight"
df_ord_dataset_disease <- data.frame(pc1, pc2, disease_bin = factor(samp$disease == "n", levels = c(T, F), labels = c("diseased", "healthy")), disease = samp$disease_simplified, study = samp$study, prev = df_ordinate$prev)

df_ord_dataset_disease %>%
    ggplot(aes(x = pc1, y = pc2, shape = disease, color = study)) +
    geom_point() +
    labs(x = "Axis 1", y = "Axis 2", title = "PCoA by dataset and disease")
```
