---
title: "Clustering of All Gut Samples using Taxonomic Abundance"
author: "Audrey Renson"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    number_sections: yes
    toc: yes
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Fetch and Cluster Gut Metagenomic Taxonomic Tables}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE)
```
# Create datasets

```{r, echo=FALSE}
## Get rid of this as soon as the function is added to the package...
curatedMetagenomicData <- function(x = "*",
           dryrun = TRUE,
           counts = FALSE,
           bugs.as.phyloseq = FALSE,
           x.is.glob = TRUE) {
    requested.datasets <- x
    all.datasets <- ls("package:curatedMetagenomicData")
    all.datasets <-
      grep("marker|gene|path|metaphlan_bugs", all.datasets, val = TRUE)
    regex <-
      ifelse(x.is.glob,
             paste(glob2rx(requested.datasets), collapse = "|"),
             requested.datasets)
    matched.datasets <- grep(regex, all.datasets, value = TRUE)
    if (dryrun) {
      message("Dry run: see return values for datasets that would be downloaded. Run with `dryrun=FALSE` to actually download these datasets.")
      return(matched.datasets)
    }
    if (!any(matched.datasets %in% all.datasets))
      stop("requested datasets do not match any available datasets.")
    eset.list <- lapply(seq_along(matched.datasets), function(i) {
      message(paste0("Working on ", matched.datasets[i]))
      eset <- do.call(get(matched.datasets[i]), list())
      if (counts) {
        exprs(eset) <-
          round(sweep(exprs(eset), 2, eset$number_reads / 100, "*"))
      }
      if(bugs.as.phyloseq && grepl("metaphlan", matched.datasets[i])){
        eset <- ExpressionSet2phyloseq(eset)
      }
      return(eset)
    })
    names(eset.list) <- matched.datasets
    return(eset.list)
  }
```

Create a list of ExpressionSet objects:
```{r, message=FALSE}
suppressPackageStartupMessages(library(curatedMetagenomicData))
eset.list <- curatedMetagenomicData("*metaphlan_bugs_list.stool", dryrun = FALSE)
```

Give them simplified titles:
```{r}
names(eset.list)
names(eset.list) <- gsub("\\..+", "", names(eset.list))
names(eset.list)
```

And add the titles to the colnames:
```{r}
for (i in 1:length(eset.list)){
    colnames(eset.list[[i]]) <- paste(names(eset.list)[[i]],
                                      colnames(eset.list[[i]]), 
                                      sep=".")
    pData(eset.list[[i]]) <- pData(eset.list[[i]])[, !sapply(pData(eset.list[[i]]), function(x) all(is.na(x)))]
    eset.list[[i]]$subjectID <- as.character(eset.list[[i]]$subjectID)
}
```
Keep leaf nodes only:
```{r}
for (i in seq_along(eset.list)){
  eset.list[[i]] <- eset.list[[i]][grep("t__", rownames(eset.list[[i]]), invert=TRUE), ]
  eset.list[[i]] <- eset.list[[i]][grep("s__|_unclassified\t", rownames(eset.list[[i]]), perl=TRUE), ]
}
```

# Merging into one big `ExpressionSet`

```{r join, message=FALSE, cache=TRUE}
joinWithRnames <- function(obj, FUN = I) {
    mylist <- lapply(obj, function(x) {
        df <- data.frame(FUN(x))
        df$rnames28591436107 <- rownames(df)
        return(df)
    })
    bigdf <- Reduce(dplyr::full_join, mylist)
    rownames(bigdf) <- make.names(bigdf$rnames28591436107)
    bigdf <- bigdf[, !grepl("^rnames28591436107$", colnames(bigdf))]
    return(bigdf)
}

pdat <- joinWithRnames(eset.list, FUN=pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN=exprs)
ab[is.na(ab)] <- 0
eset <- ExpressionSet(assayData = as.matrix(ab), 
                      phenoData = AnnotatedDataFrame(pdat))
eset
```

# Convert to `phyloseq` object

```{r}
suppressPackageStartupMessages(library(phyloseq))
source("https://raw.githubusercontent.com/waldronlab/presentations/master/Waldron_2016-06-07_EPIC/metaphlanToPhyloseq.R")
pseq <- metaphlanToPhyloseq(tax=exprs(eset), metadat=pData(eset), split=".")
```

```{r}
## Coordinated color scheme
source("https://raw.githubusercontent.com/waldronlab/curatedMetagenomicData/master/vignettes/extras/colorPalette_curatedMetagenomicData.R")

```



# Principle Coordinates Analysis (PCoA) Plots

## Colored by whether the data are from community samples or the Human Microbiome Project. 

```{r create_HMP_vs_community_covariate}
#extract sample data
samp <- data.frame(sample_data(pseq))
#calculate a covariate for HMP vs community
samp$source = factor(samp$study == "HMP_2012", levels=c(T,F), labels=c("HMP","Community"))
#update sample data with HMP covariate
sample_data(pseq) <- samp
```

```{r dissimilarity_matrices}
suppressPackageStartupMessages(library(cluster))

#dissimilarity matrices
    dist_bray <- distance(pseq, method = "bray")
    dist_js <- distance(pseq, method="jsd")
    dist_rjs <- sqrt(dist_js)
    
```

```{r PCoA_ordination}
ord_bray <- ordinate(pseq, method="PCoA", distance=dist_bray)
ord_JS  <- ordinate(pseq, method="PCoA", distance=dist_js)
ord_RJS <- ordinate(pseq, method="PCoA", distance=dist_rjs)
```

```{r plot_PCoA}
plot_ordination(pseq, ord_bray, color="source") + 
  ggplot2::ggtitle("Bray-Curtis Principal Coordinates Analysis")
plot_ordination(pseq, ord_JS, color="source") + 
  ggplot2::ggtitle("Jensen-Shannon Principal Coordinates Analysis")
plot_ordination(pseq, ord_RJS, color="source") + 
  ggplot2::ggtitle("Root Jensen-Shannon Principal Coordinates Analysis")

```


## Colored by cluster
```{r PCoA_bray234}
    
#partitioning around medoids

samp$bray_cluster_2 <- factor(pam(dist_bray, k=2, cluster.only = T))
samp$bray_cluster_3 <- factor(pam(dist_bray, k=3, cluster.only = T))
samp$bray_cluster_4 <- factor(pam(dist_bray, k=4, cluster.only = T))
sample_data(pseq) <- samp

plot_ordination(pseq, ord_bray, color="bray_cluster_2") + 
  ggplot2::ggtitle("Bray-Curtis PCoA with 2 Clusters") 
plot_ordination(pseq, ord_bray, color="bray_cluster_3") + 
  ggplot2::ggtitle("Bray-Curtis PCoA with 3 Clusters")
plot_ordination(pseq, ord_bray, color="bray_cluster_4") + 
  ggplot2::ggtitle("Bray-Curtis PCoA with 4 Clusters") 
```

## Colored by Prevotella copri
```{r PCoA_Prevotella}

Prev <- as.numeric(otu_table(pseq)["s__Prevotella_copri",])
samp$Prevotella_copri <- Prev
sample_data(pseq) <- samp
plot_ordination(pseq, ord_bray, color="Prevotella_copri")
```


# Cluster Validation
## 1. Prediction Strength
```{r calculate_Prediction_strength}


### LEGACY Source code from github for prediction strength allowing distance matrices
# suppressPackageStartupMessages(library(RCurl))
# string = getURL("https://raw.githubusercontent.com/danknights/mice8992-2016/master/src/prediction.strength.r")
# eval(parse(text=string))

suppressPackageStartupMessages(library(fpc))

ps_bray <- prediction.strength(dist_bray, Gmin = 2, Gmax = 10, clustermethod = pamkCBI)
ps_js <- prediction.strength(dist_js, Gmin = 2, Gmax = 10, clustermethod = pamkCBI)
ps_rjs <- prediction.strength(dist_rjs,Gmin = 2, Gmax = 10, clustermethod = pamkCBI)
```
```{r plot_function}
#function to plot cluster validation measures
plot_cluster_validation = function(bray,js,rjs, legend=T,...) {
  plot(2:10, bray, type="b", pch=1, xlab="Number of Clusters", ...)
  lines(2:10, js, type="b", pch=2, lty=2)
  lines(2:10, rjs, type="b", pch=22, lty=3)
  if(legend) legend("topright", legend = c("Bray-Curtis","Jensen-Shannon", "Root Jensen-Shannon"), pch=c(1,2,22), lty=1:3)
}

```
```{r plot_Prediction_Strength}
plot_cluster_validation(ps_bray$mean.pred[2:10], ps_js$mean.pred[2:10], ps_rjs$mean.pred[2:10], ylab="Prediction Strength",ylim=c(0,1.1),legend=F)
abline(.9,0, lty=5, col="grey70")
abline(0.8,0,lty=8, col="grey70")
text("Strong support", x=9,y=1, col="grey70")
text("Moderate support", x=9, y=.85, col="grey70")
text("Little or no support", x=9, y=.6, col="grey70")
```




## 2. Silhouette Index
```{r Silhouette_Index}
suppressPackageStartupMessages(library(cluster))

#get clusters for k=2:10
   pam_bray = sapply(2:10, function(i) pam(dist_bray, k=i, cluster.only = T))
   pam_js = sapply(2:10, function(i) pam(dist_js, k=i, cluster.only = T))
   pam_rjs = sapply(2:10, function(i) pam(dist_rjs, k=i, cluster.only = T))

#silhouette indices for k=2:10, distance = {bray-curtis, jenson-shannon, root jenson-shannon}
   
   si_bray = apply(pam_bray, 2, function(i) mean(silhouette(i, dist_bray)[,3]))
   si_js = apply(pam_js, 2, function(i) mean(silhouette(i, dist_js)[,3]))
   si_rjs = apply(pam_rjs, 2, function(i) mean(silhouette(i, dist_rjs)[,3]))
```

```{r plot_Silhouette}
#plot silhouette widths
   plot_cluster_validation(si_bray,si_js,si_rjs, legend=F, ylab="Average silhouette width", ylim=c(0,1))

abline(.75,0, lty=5, col="grey70")
abline(0.5,0,lty=5, col="grey70")
abline(0.25,0,lty=5, col="grey70")
text("Strong support", x=9,y=.8, col="grey70")
text("Moderate support", x=9, y=.6, col="grey70")
text("Weak and could be artificial", x=8.5, y=.4, col="grey70")
text("No substantial structure", x=8.5, y=.2, col="grey70")
```

## Calinski-Harabasz
```{r Calinski-Harabasz}
suppressPackageStartupMessages(library(fpc))

ch_bray = apply(pam_bray, 2, function(i) cluster.stats(dist_bray, i)$ch)
ch_js = apply(pam_js, 2, function(i) cluster.stats(dist_js, i)$ch)
ch_rjs = apply(pam_rjs, 2, function(i) cluster.stats(dist_rjs, i)$ch)

```
```{r plot_Calinski}
#plot calinski-harabasz
plot_cluster_validation(ch_bray, ch_js, ch_rjs, legend=T, ylab="Calinski-Harabasz score", ylim=c(0,300))
```

## PCoA k=2 & gradient for Prevotella & Bacteroides
```{r PCoA_prevotella_bacteroides}
#extract 1st 2 principle vectors
pc1 <- ord_bray$vectors[,1]
pc2 <- ord_bray$vectors[,2]

#calculate sum of bacteroides species for each sample
otu_tax <- attr(otu_table(pseq), "dimnames")[[1]]
otu_bacteroides <- otu_table(pseq)[grep("s__Bacteroides", otu_tax),]
sum_bacteroides <- apply(otu_bacteroides, 2, sum)

#create data.frame for ordination plot
df_ordinate <- data.frame(pc1, pc2, bact = sum_bacteroides, prev = Prev, bray2 = as.numeric(samp$bray_cluster_2)+20)
df_bact <- df_ordinate[df_ordinate$bray2==21,]
df_prev <- df_ordinate[df_ordinate$bray2==22,]

#plot
suppressPackageStartupMessages(library(ggplot2))

p <- ggplot() 
#Prevotella
p <- p +  geom_point(data= df_prev, aes(x=pc1, y=pc2, shape=factor(bray2), fill=prev), shape=21, size=4)
p <- p + scale_fill_gradient(low="white", high=purple,
          guide = guide_colorbar(title="Prevotella copri \n(cluster 2)"))
#Bacteroides
p <- p + geom_point(data=df_bact, aes(x=pc1, y=pc2, shape=factor(bray2), color=bact), shape=22, size=4)
p <- p + scale_color_gradient(low=gray, high=blueGreen, 
              guide=guide_colorbar(title="Bacteroides \n(cluster 1)"))
#Labels
p <- p + labs(x="Axis 1",y="Axis 2", title="PCoA by selected abundances, displaying 2 clusters")
p



```

```{r PCoA_dataset_diseasestate, fig.width=11}
#some data cleaning for PCoA plot by disease and dataset
samp$disease[samp$disease %in% c("obesity","obese")] <- "obesity"
samp$disease[samp$disease %in% c("underweight","leaness")] <- "underweight"

samp$disease_simplified <- NA
samp$disease_simplified[samp$disease == "cancer"] <- "cancer"
samp$disease_simplified[samp$disease %in% c("small_adenoma","large_adenoma")] <- "adenoma"
samp$disease_simplified[samp$disease == "cirrhosis"] <- "cirrhosis"
samp$disease_simplified[samp$disease %in% c("t2d","impaired_glucose_tolerance")] <- "t2d / impaired glucose tolerance"
samp$disease_simplified[samp$disease %in% c("ibd_crohn_disease","ibd_ulcerative_colitis")] <- "ibd"
samp$disease_simplified[samp$disease %in% c("obesity","overweight")] <- "obese or overweight"
#samp$disease_simplified[samp$disease %in% c("n","n_relative")] <- "n"
 

df_ord_dataset_disease <- data.frame(pc1, pc2, 
                                     disease_bin=factor(samp$disease=="n", 
                                                    levels=c(T,F), 
                                              labels=c("diseased","healthy")),
                                     disease = samp$disease_simplified,
                                     study=samp$study,
                                     prev=df_ordinate$prev)

# p <- ggplot(data=df_ord_dataset_disease, aes(x=pc1, y=pc2, shape=disease,
#                                              color=study, size=prev)) 
# p <- p + geom_point()
# p <- p + labs(x="Axis 1", y="Axis 2", title="PCoA by dataset and disease")
# p <- p + scale_shape_manual(values = 19:25)
# p <- p + guides(color=guide_legend(ncol=2), shape=guide_legend(ncol=2), size=guide_legend(ncol=2, title="Prevotella copri"))
# p
p <- ggplot(data=df_ord_dataset_disease, aes(x=pc1, y=pc2, shape=disease,
                                             color=study)) 
p <- p + geom_point()
p <- p + labs(x="Axis 1", y="Axis 2", title="PCoA by dataset and disease")
p
```




