---
title: "Paper Figures"
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{Paper Figures}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.show = "hold")
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

```{r libraries, include=FALSE}
library(curatedMetagenomicData)
library(randomForest)
library(caret)
library(pROC)
library(ggplot2)
library(cluster)
library(fpc)
library(RColorBrewer)
library(stats)
library(reshape2)
library(utils)
```

# Coordinated Color Scheme

```{r color_scheme}
blue <- "#3366aa"
blueGreen <- "#11aa99"
green <- "#66aa55"
paleYellow <- "#cccc55"
yellow <- "#ffee33"
gray <- "#777777"
purple <- "#992288"
red <- "#ee3333"
orange <- "#ee7722"
black <- "#000000"
brown <- "#655643"
white <- "#ffffff"
pallet <- c(purple, blue, blueGreen, green, paleYellow, yellow, orange, red, 
            white, gray, brown, black)
n <- length(pallet)
image(1:n, 1, as.matrix(1:n), col = pallet, xlab = "", ylab = "", xaxt = "n", 
      yaxt = "n", bty = "n")
```

# Figure 1, Example 1: Classification

```{r}
dataset_list <- c("KarlssonFH_2013 (T2D)", "LeChatelierE_2013 (Obesity)", "NielsenHB_2014 (IBD)", "QinJ_2012 (T2D)", "QinN_2014 (Cirrhosis)", "ZellerG_2014 (CRC)")
class_list <- c("t2d", "obesity", "ibd", "t2d", "cirrhosis", "cancer")
data_list <- matrix(nrow = 5, ncol = length(dataset_list))
data_list[1, ] <- c("EH277", "EH283", "EH301", "EH319", "EH325", "EH361")  # Species abundance
data_list[2, ] <- c("EH278", "EH284", "EH302", "EH320", "EH326", "EH362")  # Pathway abundance
data_list[3, ] <- c("EH279", "EH285", "EH303", "EH321", "EH327", "EH363")  # Pathway coverage
data_list[4, ] <- c("EH275", "EH281", "EH299", "EH317", "EH323", "EH359")  # Maker abundance
data_list[5, ] <- c("EH276", "EH282", "EH300", "EH318", "EH324", "EH360")  # Marker presence

eh <- ExperimentHub()

for (j in 1:length(data_list[, 1])) {
    for (i in 1:length(dataset_list)) {
        taxabund <- eh[[data_list[j, i]]]
        
        feat <- t(exprs(taxabund))
        if (j == 1) {
            feat <- feat[, grep("(s__|unclassified)", colnames(feat))]
            feat <- feat[, -grep("t__", colnames(feat))]
        }
        meta <- pData(taxabund)["disease"]
        all <- cbind(feat, meta)
        if (i == 1) {
            all <- subset(all, disease != "impaired_glucose_tolerance")
        }
        if (i == 2) {
            all <- subset(all, disease != "n")
        }
        if (i == 3) {
            all$disease[all$disease == "ibd_ulcerative_colitis"] <- "ibd"
            all$disease[all$disease == "ibd_crohn_disease"] <- "ibd"
            all$disease[all$disease == "n_relative"] <- "n"
        }
        if (i == 4) {
            all <- subset(all, disease != "na")
        }
        if (i == 6) {
            all <- subset(all, disease != "large_adenoma")
            all$disease[all$disease == "small_adenoma"] <- "n"
        }
        
        assign(paste("rf", j, i, sep = "_"), train(disease ~ ., data = all, preProc = c("zv", "scale", "center"), method = "rf", ntree = 500, tuneGrid = expand.grid(.mtry = c(10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 125, 150, 175, 200, 300, 400, 500)), trControl = trainControl(method = "repeatedcv", number = 10, search = "grid", summaryFunction = twoClassSummary, classProbs = TRUE, savePredictions = TRUE)))
    }
}

ci_bugs <- c()
ci_pab <- c()
ci_pcov <- c()
ci_mab <- c()
ci_mpr <- c()

for (i in 1:length(dataset_list)) {
    rf <- get(paste("rf", 1, i, sep = "_"))
    ci_bugs <- c(ci_bugs, auc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry]))
    
    rf <- get(paste("rf", 2, i, sep = "_"))
    ci_pab <- c(ci_pab, auc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry]))
    
    rf <- get(paste("rf", 3, i, sep = "_"))
    ci_pcov <- c(ci_pcov, auc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry]))
    
    rf <- get(paste("rf", 4, i, sep = "_"))
    ci_mab <- c(ci_mab, auc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry]))
    
    rf <- get(paste("rf", 5, i, sep = "_"))
    ci_mpr <- c(ci_mpr, auc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry]))
}

ci <- data.frame(ci_bugs, ci_pab, ci_pcov, ci_mab, ci_mpr)
colnames(ci) <- c("Taxonomic abundance", "Pathway abundance", "Pathway coverage", "Marker abundance", "Marker presence")
ci.r <- abs(cor(ci))
ci.col <- dmat.color(ci.r)
ci.o <- order.single(ci.r)

par(oma = c(4, 1, 1, 1))

cpairs(ci, ci.o, panel.colors = ci.col, col = alpha(pallet, 0.75), pch = 16, cex = 2, gap = 0.5, main = "AUC")

par(fig = c(0, 1, 0, 1), oma = c(0, 0, 0, 0), mar = c(0, 0, 0, 0), new = TRUE)
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n")
legend("bottom", xpd = TRUE, horiz = TRUE, inset = c(0, 0), bty = "n", pch = 16, cex = 0.6, dataset_list, col = pallet)
```

# Figure 1, Example 2: Clustering

```{r}
ggplot() +
    geom_point(data = df_prev, aes(x = pc1, y = pc2, shape = factor(bray2), fill = prev), shape = 21, size = 4) +
    scale_fill_gradient(low = "white", high = purple, guide = guide_colorbar(title = "Prevotella copri \n(cluster 2)")) +
    geom_point(data = df_bact, aes(x = pc1, y = pc2, shape = factor(bray2), color = bact), shape = 22, size = 4) +
    scale_color_gradient(low = gray, high = blueGreen, guide = guide_colorbar(title = "Bacteroides \n(cluster 1)")) +
    labs(x = "Axis 1", y = "Axis 2", title = paste0("PCoA by selected abundances, displaying 2 clusters (n=", nrow(df_ordinate), ")"))
```

# Figure 1, Example 3: Abundance across samples

```{r}

```

# Figure 1, Example 4: Species-pathway correlation

```{r}

```

# Supplemental Figure 1: Health status classification from species abundance.
```{r}

```

# Supplemental Figure 2: PCoA plots colored for dataset + disease states.
```{r}

```

# Supplemental Figure 3. Clustering scores for enterotypes in gut WGS samples.
```{r}

```

# Supplemental Figure 4: Top correlations between metabolic pathways and genera.
```{r}

```

# Supplemental Figure 5: Alpha diversity of taxa from 11 studies of the gut microbiome.
```{r}

```

# Session Info
```{r}
sessionInfo()
```
