---
title: "Paper Figures"
abstract: >
    The following analysis represents all the code that is necessary to 
    reproduce the results shown in Pasolli, E., Schiffer, L., Renson, A., 
    Obenchain, V., Manghi, P., Truong, D. T., Dowd, JB, Huttenhower, C, 
    Morgan, M, Segata, N,  Waldron, L. (2017). Accessible, curated metagenomic 
    data through ExperimentHub. bioRxiv. doi&#58;10.1101/103085. Please note 
    that much of this analysis occurred during the development stages of 
    curatedMetagenomicData and was written by multiple authors. As such, while 
    this analysis is methodologically sound, there may now exist more efficient 
    ways to produce the same results syntactically.
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{Paper Figures}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.show = "hold")
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

```{r libraries, include=FALSE, cache=FALSE}
library(curatedMetagenomicData)
library(randomForest)
library(caret)
library(pROC)
library(ggplot2)
library(cluster)
library(fpc)
library(RColorBrewer)
library(stats)
library(reshape2)
library(utils)
library(phyloseq)
library(cluster)
library(dplyr)
```

# Coordinated Color Scheme

```{r colors}
blue <- "#3366aa"
blueGreen <- "#11aa99"
green <- "#66aa55"
paleYellow <- "#cccc55"
gray <- "#e5e5e5"
purple <- "#992288"
red <- "#ee3333"
orange <- "#ee7722"
yellow <- "#ffee33"
darkGray <- "#b3b3b3"
brown <- "#655643"
lightBlack <- "#777777"
black <- "#000000"
pallet <- c(blue, blueGreen, green, paleYellow, gray, purple, red, orange, 
            yellow, darkGray, brown, lightBlack, black)
n <- length(pallet)
image(1:n, 1, as.matrix(1:n), col = pallet, xlab = "", ylab = "", xaxt = "n", 
      yaxt = "n", bty = "n")
```

# Figure 1, Example 1: Classification

```{r fig1eg1, fig.height=8, fig.width=8, eval=FALSE}
dataset_list <-
    c(
        "KarlssonFH_2013 (T2D)",
        "LeChatelierE_2013 (Obesity)",
        "NielsenHB_2014 (IBD)",
        "QinJ_2012 (T2D)",
        "QinN_2014 (Cirrhosis)",
        "ZellerG_2014 (CRC)"
    )
class_list <- c("t2d", "obesity", "ibd", "t2d", "cirrhosis", "cancer")
data_list <- matrix(nrow = 5, ncol = length(dataset_list))
## update
data_list[1, ] <- c("EH277", "EH283", "EH301", "EH319", "EH325", "EH361")  # Species abundance
data_list[2, ] <- c("EH278", "EH284", "EH302", "EH320", "EH326", "EH362")  # Pathway abundance
data_list[3, ] <- c("EH279", "EH285", "EH303", "EH321", "EH327", "EH363")  # Pathway coverage
data_list[4, ] <- c("EH275", "EH281", "EH299", "EH317", "EH323", "EH359")  # Maker abundance
data_list[5, ] <- c("EH276", "EH282", "EH300", "EH318", "EH324", "EH360")  # Marker presence

## update
eh <- ExperimentHub()

for (j in 1:length(data_list[, 1])) {
    for (i in 1:length(dataset_list)) {
        print(dataset_list[i])
        taxabund <- eh[[data_list[j, i]]]
        
        feat <- t(exprs(taxabund))
        if (j == 1) {
            feat <- feat[, grepl("(s__|unclassified)", colnames(feat))]
            feat <- feat[, !grepl("t__", colnames(feat))]
        }
        meta <- pData(taxabund)["disease"]
        all <- cbind(feat, meta)
        if (i == 1) {
            print("disease != impaired_glucose_tolerance")
            all <- subset(all, disease != "impaired_glucose_tolerance")
        }
        if (i == 2) {
            print("disease != n")
            all <- subset(all, disease != "n")
        }
        if (i == 3) {
            print("disease ibd")
            all$disease[all$disease == "ibd_ulcerative_colitis"] <- "ibd"
            all$disease[all$disease == "ibd_crohn_disease"] <- "ibd"
            all$disease[all$disease == "n_relative"] <- "n"
        }
        if (i == 4) {
            print("disease != na")
            all <- subset(all, disease != "na")
        }
        if (i == 6) {
            print("disease large_adenoma")
            all <- subset(all, disease != "large_adenoma")
            all$disease[all$disease == "small_adenoma"] <- "n"
        }
        
        assign(
            paste("rf", j, i, sep = "_"),
            train(
                disease ~ .,
                data = all,
                preProc = c("zv", "scale", "center"),
                method = "rf",
                ntree = 500,
                tuneGrid = expand.grid(
                    .mtry = c(10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 125, 150, 175, 200, 300, 400, 500)
                ),
                trControl = trainControl(
                    method = "repeatedcv",
                    number = 10,
                    search = "grid",
                    summaryFunction = twoClassSummary,
                    classProbs = TRUE,
                    savePredictions = TRUE
                )
            )
        )
    }
}

ci_bugs <- c()
ci_pab <- c()
ci_pcov <- c()
ci_mab <- c()
ci_mpr <- c()

for (i in 1:length(dataset_list)) {
    rf <- get(paste("rf", 1, i, sep = "_"))
    ci_bugs <- c(ci_bugs, auc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry]))
    
    rf <- get(paste("rf", 2, i, sep = "_"))
    ci_pab <- c(ci_pab, auc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry]))
    
    rf <- get(paste("rf", 3, i, sep = "_"))
    ci_pcov <- c(ci_pcov, auc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry]))
    
    rf <- get(paste("rf", 4, i, sep = "_"))
    ci_mab <- c(ci_mab, auc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry]))
    
    rf <- get(paste("rf", 5, i, sep = "_"))
    ci_mpr <- c(ci_mpr, auc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry]))
}

ci <- data.frame(ci_bugs, ci_pab, ci_pcov, ci_mab, ci_mpr)
colnames(ci) <- c("Taxonomic abundance", "Pathway abundance", "Pathway coverage", "Marker abundance", "Marker presence")
ci.r <- abs(cor(ci))
ci.col <- dmat.color(ci.r)
ci.o <- order.single(ci.r)

par(oma = c(4, 1, 1, 1))

cpairs(ci, ci.o, panel.colors = ci.col, col = alpha(pallet, 0.75), pch = 16, cex = 2, gap = 0.5, main = "AUC")

par(fig = c(0, 1, 0, 1), oma = c(0, 0, 0, 0), mar = c(0, 0, 0, 0), new = TRUE)
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n")
legend("bottom", xpd = TRUE, horiz = TRUE, inset = c(0, 0), bty = "n", pch = 16, cex = 0.6, dataset_list, col = pallet)
```

# Figure 1, Example 2: Clustering

```{r fig1eg2, fig.height=8, fig.width=8}
## update
eh <- ExperimentHub()
myquery <- query(eh, "curatedMetagenomicData")

myquery.stool <- myquery[grepl("stool", myquery$title) & grepl("bugs", myquery$title), ]

eset.list <- lapply(names(myquery.stool), function(x) myquery.stool[[x]])

names(eset.list) <- myquery.stool$title
names(eset.list) <- gsub("\\..+", "", myquery.stool$title)

for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <- paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <- pData(eset.list[[i]])[, !sapply(pData(eset.list[[i]]), function(x) all(is.na(x)))]
    eset.list[[i]]$subjectID <- as.character(eset.list[[i]]$subjectID)
}

for (i in seq_along(eset.list)) {
    eset.list[[i]] <- eset.list[[i]][grep("t__", rownames(eset.list[[i]]), invert = TRUE), ]
    eset.list[[i]] <- eset.list[[i]][grep("s__|_unclassified\t", rownames(eset.list[[i]]), perl = TRUE), ]
}

joinWithRnames <- function(obj, FUN = I) {
    mylist <- lapply(obj, function(x) {
        df <- data.frame(FUN(x))
        df$rnames28591436107 <- rownames(df)
        return(df)
    })
    bigdf <- Reduce(full_join, mylist)
    rownames(bigdf) <- make.names(bigdf$rnames28591436107)
    bigdf <- bigdf[, !grepl("^rnames28591436107$", colnames(bigdf))]
    return(bigdf)
}

pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0
eset <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))

## update
source("https://raw.githubusercontent.com/waldronlab/presentations/master/Waldron_2016-06-07_EPIC/metaphlanToPhyloseq.R")
pseq <- metaphlanToPhyloseq(tax = exprs(eset), metadat = pData(eset), split = ".")

samp <- data.frame(sample_data(pseq))

dist_bray <- distance(pseq, method = "bray")

ord_bray <- ordinate(pseq, method = "PCoA", distance = dist_bray)

samp$bray_cluster_2 <- factor(pam(dist_bray, k = 2, cluster.only = TRUE))
sample_data(pseq) <- samp

Prev <- as.numeric(otu_table(pseq)["s__Prevotella_copri", ])
samp$Prevotella_copri <- Prev
sample_data(pseq) <- samp

pc1 <- ord_bray$vectors[, 1]
pc2 <- ord_bray$vectors[, 2]

otu_tax <- attr(otu_table(pseq), "dimnames")[[1]]
otu_bacteroides <- otu_table(pseq)[grep("s__Bacteroides", otu_tax), ]
sum_bacteroides <- apply(otu_bacteroides, 2, sum)

df_ordinate <- data.frame(pc1, pc2, bact = sum_bacteroides, prev = Prev, bray2 = as.numeric(samp$bray_cluster_2) + 20)
df_bact <- df_ordinate[df_ordinate$bray2 == 21, ]
df_prev <- df_ordinate[df_ordinate$bray2 == 22, ]

ggplot() +
    geom_point(data = df_prev, aes(x = pc1, y = pc2, shape = factor(bray2), fill = prev), shape = 21, size = 4) +
    scale_fill_gradient(low = "white", high = purple, guide = guide_colorbar(direction = "horizontal", title = "Prevotella copri \n(cluster 2)")) +
    geom_point(data = df_bact, aes(x = pc1, y = pc2, shape = factor(bray2), color = bact), shape = 22, size = 4) +
    scale_color_gradient(low = lightBlack, high = blueGreen, guide = guide_colorbar(direction = "horizontal", title = "Bacteroides \n(cluster 1)")) +
    labs(x = "Axis 1", y = "Axis 2", title = "PCoA on species abundance, displaying 2 clusters") +
    theme(axis.ticks = element_blank(), axis.text = element_blank(), legend.box = "vertical", legend.position = c(0.8, 0.1), plot.title = element_text(hjust = 0.5))
```

# Figure 1, Example 3: Abundance across samples

```{r fig1eg3, fig.height=8, fig.width=12}
## update
eh <- ExperimentHub()
myquery <- query(eh, "curatedMetagenomicData")

myquery.stool <- myquery[grepl("stool", myquery$title) & grepl("pathabundance", myquery$title), ]

eset.list <- lapply(names(myquery.stool), function(x) myquery.stool[[x]])

names(eset.list) <- myquery.stool$title
names(eset.list) <- gsub("\\..+", "", myquery.stool$title)

for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <- paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <- pData(eset.list[[i]])[, !sapply(pData(eset.list[[i]]), function(x) all(is.na(x)))]
    eset.list[[i]]$subjectID <- as.character(eset.list[[i]]$subjectID)
}

for (i in seq_along(eset.list)) {
    eset.list[[i]] <- eset.list[[i]][!grepl("\\|", rownames(eset.list[[i]])), ]
}

pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0
eset_pathway <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))

myquery.stool <- myquery[grepl("stool", myquery$title) & grepl("bugs", myquery$title), ]

eset.list <- lapply(names(myquery.stool), function(x) myquery.stool[[x]])

names(eset.list) <- myquery.stool$title
names(eset.list) <- gsub("\\..+", "", myquery.stool$title)

for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <-
        paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <-
        pData(eset.list[[i]])[,!sapply(pData(eset.list[[i]]), function(x)
            all(is.na(x)))]
    eset.list[[i]]$subjectID <-
        as.character(eset.list[[i]]$subjectID)
}

for (i in seq_along(eset.list)) {
    eset.list[[i]] <-
        eset.list[[i]][grep("t__", rownames(eset.list[[i]]), invert = TRUE),]
    eset.list[[i]] <-
        eset.list[[i]][grep("s__|_unclassified\t", rownames(eset.list[[i]]), perl = TRUE),]
}

pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0
eset_bugs <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))

## update
pseq <- metaphlanToPhyloseq(tax = exprs(eset_bugs), metadat = pData(eset_bugs), split = ".")

glom <- tax_glom(pseq, taxrank = "Phylum")

top8phyla <- names(sort(taxa_sums(glom), TRUE)[1:8])
phyla8_subset <- prune_taxa(top8phyla, glom)

phyla_to_sort <-
    data.frame(
        id = 1:8,
        phyla = as.character(tax_table(phyla8_subset)[, "Phylum"]),
        otu = as.character(taxa_names(phyla8_subset))
    )
rownames(phyla_to_sort) <- phyla_to_sort$otu

phylum_ranks <-
    phyla8_subset %>%
    otu_table %>%
    rowSums %>%
    sort(TRUE) %>%
    names

phyla_to_sort <- phyla_to_sort[phylum_ranks, ]

prop <- transform_sample_counts(phyla8_subset, function(i) i/sum(i))

bardat <-
    psmelt(prop) %>%
    select(OTU, Sample, Abundance) %>%
    mutate(Sample = as.numeric(factor(Sample)), OTU = factor(OTU, levels = phyla_to_sort$otu, labels = phyla_to_sort$phyla))

firmicutes_order <-
    bardat %>%
    filter(OTU == "Firmicutes") %>%
    arrange(Abundance) %>%
    select(Sample)

bardat %<>%
    mutate(Sample = as.numeric(factor(Sample, levels = factor(firmicutes_order$Sample)))) %>%
    arrange(desc(OTU), Abundance)

set.seed(14)

bardat %>%
    ggplot(aes(x = Sample, y = Abundance, fill = OTU)) +
    geom_area() +
    scale_fill_manual(values = sample(c(blue, yellow, green, gray, purple, 
                                        red, orange, paleYellow), 
                                      size = 8, replace = FALSE), 
                      guide = guide_legend(nrow = 4, ncol = 2)) +
    theme(axis.text.x = element_blank(), axis.title.x = element_blank(), axis.ticks.x = element_blank(), legend.position = c(0.8, 0.8), legend.direction = "vertical", legend.title = element_blank(), panel.background = element_blank())
```

# Figure 1, Example 4: Species-pathway correlation

```{r fig1eg4, fig.height=8, fig.width=8}
eset_pathway$prev <- as.numeric(exprs(eset_bugs)[grep("s__Prevotella_copri", rownames(exprs(eset_bugs))), ])

cor_est_p <- function(x1, x2) {
    cor <- cor.test(x1, x2)
    c(r = cor$estimate, p = cor$p.value)
}

cors <- t(sapply(featureNames(eset_pathway), function(i) cor_est_p(exprs(eset_pathway)[i, ], eset_pathway$prev)))

feature <- rownames(cors)

cors <- as.data.frame(cors)
cors$feature <- feature
cors <- na.omit(cors)

par(mar = c(10, 10, 10, 10))
qplot(
    x = eset_pathway$prev,
    y = exprs(eset_pathway)[cors$feature[cors$r.cor == max(cors$r.cor)],],
    xlab = "Prevotella copri abundance",
    ylab = "Pathway abundance",
    main = "Inosine 5 phosphate biosynthesis ",
    colour = I(black),
    shape = I(1)
) +
    theme(plot.title = element_text(hjust = 0.5))
```

# Supplemental Figure 1: Health status classification from species abundance.

```{r suppfig1, fig.height=8, fig.width=8}
dataset_list <-
    c(
        "KarlssonFH_2013 (T2D)",
        "LeChatelierE_2013 (Obesity)",
        "NielsenHB_2014 (IBD)",
        "QinJ_2012 (T2D)",
        "QinN_2014 (Cirrhosis)",
        "ZellerG_2014 (CRC)"
    )
class_list <- c("t2d", "obesity", "ibd", "t2d", "cirrhosis", "cancer")
data_list <- c("EH277", "EH283", "EH301", "EH319", "EH325", "EH361")

eh <- ExperimentHub()

for (i in 1:length(dataset_list)) {
    taxabund <- eh[[data_list[i]]]
    
    feat <- t(exprs(taxabund))
    feat <- feat[, grep("(s__|unclassified)", colnames(feat))]
    feat <- feat[, -grep("t__", colnames(feat))]
    meta <- pData(taxabund)["disease"]
    all <- cbind(feat, meta)
    if (i == 1) {
        all <- subset(all, disease != "impaired_glucose_tolerance")
    }
    if (i == 2) {
        all <- subset(all, disease != "n")
    }
    if (i == 3) {
        all$disease[all$disease == "ibd_ulcerative_colitis"] <- "ibd"
        all$disease[all$disease == "ibd_crohn_disease"] <- "ibd"
        all$disease[all$disease == "n_relative"] <- "n"
    }
    if (i == 4) {
        all <- subset(all, disease != "na")
    }
    if (i == 6) {
        all <- subset(all, disease != "large_adenoma")
        all$disease[all$disease == "small_adenoma"] <- "n"
    }
    
    assign(
        paste("rf", i, sep = "_"),
        train(
            disease ~ .,
            data = all,
            preProc = c("zv", "scale", "center"),
            method = "rf",
            ntree = 500,
            tuneGrid = expand.grid(
                .mtry = c(10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 125, 150, 175, 200, 300, 400, 500)
            ),
            trControl = trainControl(
                method = "repeatedcv",
                number = 10,
                search = "grid",
                summaryFunction = twoClassSummary,
                classProbs = TRUE,
                savePredictions = TRUE
            )
        )
    )
}

pallet_reduced <- c(blue, green, gray, purple, red, black)

for (i in 1:length(dataset_list)) {
    rf <- get(paste("rf", i, sep = "_"))
    
    if (i == 1) {
        plot.roc(
            rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry],
            rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry],
            grid = TRUE,
            ci = TRUE,
            xaxs = "i",
            yaxs = "i",
            col = pallet_reduced[i],
            lty = 1
        )
    } else {
        plot.roc(
            rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry],
            rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry],
            grid = TRUE,
            ci = TRUE,
            xaxs = "i",
            yaxs = "i",
            col = pallet_reduced[i],
            lty = 1,
            add = TRUE
        )
    }
}

legend(
    "bottomright",
    box.lwd = 0,
    box.col = "transparent",
    bg = "transparent",
    lwd = 2,
    legend = dataset_list,
    col = pallet_reduced
    
)
```

# Supplemental Figure 2: PCoA plots colored for dataset + disease states.

```{r suppfig2, fig.height=8, fig.width=12}
eset.list <- curatedMetagenomicData("*metaphlan_bugs_list.stool", dryrun = FALSE)

names(eset.list) <- gsub("\\..+", "", names(eset.list))

for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <-
        paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <-
        pData(eset.list[[i]])[,!sapply(pData(eset.list[[i]]), function(x)
            all(is.na(x)))]
    eset.list[[i]]$subjectID <-
        as.character(eset.list[[i]]$subjectID)
}

for (i in seq_along(eset.list)) {
    eset.list[[i]] <-
        eset.list[[i]][grep("t__", rownames(eset.list[[i]]), invert = TRUE),]
    eset.list[[i]] <-
        eset.list[[i]][grep("s__|_unclassified\t", rownames(eset.list[[i]]), perl = TRUE),]
}

pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))

ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0

eset <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))

## update
metaphlanToPhyloseq <- function(tax, metadat = NULL, simplenames = TRUE, roundtointeger = FALSE, split = "|") {
    xnames <- rownames(tax)
    shortnames <- gsub(paste0(".+\\", split), "", xnames)
    if (simplenames) {
        rownames(tax) <- shortnames
    }
    if (roundtointeger) {
        tax <- round(tax * 10000)
    }
    x2 <- strsplit(xnames, split = split, fixed = TRUE)
    taxmat <- matrix(NA, ncol = max(sapply(x2, length)), nrow = length(x2))
    colnames(taxmat) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "Strain")[1:ncol(taxmat)]
    rownames(taxmat) <- rownames(tax)
    for (i in 1:nrow(taxmat)) {
        taxmat[i, 1:length(x2[[i]])] <- x2[[i]]
    }
    taxmat <- gsub("[a-z]__", "", taxmat)
    taxmat <- tax_table(taxmat)
    otutab <- otu_table(tax, taxa_are_rows = TRUE)
    if (is.null(metadat)) {
        res <- phyloseq(taxmat, otutab)
    } else {
        res <- phyloseq(taxmat, otutab, sample_data(metadat))
    }
    return(res)
}

## update
pseq <- metaphlanToPhyloseq(tax = exprs(eset), metadat = pData(eset), split = ".")

samp <- data.frame(sample_data(pseq))
samp$source <- factor(samp$study == "HMP_2012", levels = c(T, F), labels = c("HMP", "Community"))
sample_data(pseq) <- samp

dist_bray <- distance(pseq, method = "bray")
dist_js <- distance(pseq, method = "jsd")
dist_rjs <- sqrt(dist_js)

ord_bray <- ordinate(pseq, method = "PCoA", distance = dist_bray)
ord_JS <- ordinate(pseq, method = "PCoA", distance = dist_js)
ord_RJS <- ordinate(pseq, method = "PCoA", distance = dist_rjs)

samp$bray_cluster_2 <- factor(pam(dist_bray, k = 2, cluster.only = T))
samp$bray_cluster_3 <- factor(pam(dist_bray, k = 3, cluster.only = T))
samp$bray_cluster_4 <- factor(pam(dist_bray, k = 4, cluster.only = T))
sample_data(pseq) <- samp

Prev <- as.numeric(otu_table(pseq)["s__Prevotella_copri", ])
samp$Prevotella_copri <- Prev
sample_data(pseq) <- samp

pc1 <- ord_bray$vectors[, 1]
pc2 <- ord_bray$vectors[, 2]

otu_tax <- attr(otu_table(pseq), "dimnames")[[1]]
otu_bacteroides <- otu_table(pseq)[grep("s__Bacteroides", otu_tax), ]
sum_bacteroides <- apply(otu_bacteroides, 2, sum)

df_ordinate <- data.frame(pc1, pc2, bact = sum_bacteroides, prev = Prev, bray2 = as.numeric(samp$bray_cluster_2) + 20)
df_bact <- df_ordinate[df_ordinate$bray2 == 21, ]
df_prev <- df_ordinate[df_ordinate$bray2 == 22, ]

samp$disease[samp$disease %in% c("obesity", "obese")] <- "obesity"
samp$disease[samp$disease %in% c("underweight", "leaness")] <- "underweight"
samp$disease_simplified[samp$disease == "cancer"] <- "cancer"
samp$disease_simplified[samp$disease %in% c("small_adenoma", "large_adenoma")] <- "adenoma"
samp$disease_simplified[samp$disease == "cirrhosis"] <- "cirrhosis"
samp$disease_simplified[samp$disease %in% c("t2d", "impaired_glucose_tolerance")] <- "t2d / impaired glucose tolerance"
samp$disease_simplified[samp$disease %in% c("ibd_crohn_disease", "ibd_ulcerative_colitis")] <- "ibd"
samp$disease_simplified[samp$disease %in% c("obesity", "overweight")] <- "obese or overweight"
df_ord_dataset_disease <-
    data.frame(
        pc1,
        pc2,
        disease_bin = factor(
            samp$disease == "n",
            levels = c(T, F),
            labels = c("diseased", "healthy")
        ),
        disease = samp$disease_simplified,
        study = samp$study,
        prev = df_ordinate$prev
    )

df_ord_dataset_disease %>%
    ggplot(aes(
        x = pc1,
        y = pc2,
        shape = disease,
        color = study,
        size = prev
    )) +
    geom_point() +
    labs(x = "Axis 1", y = "Axis 2", title = "PCoA by dataset and disease") +
    scale_shape_manual(values = 19:25) +
    guides(
        color = guide_legend(ncol = 2, title = "Study"),
        shape = guide_legend(ncol = 2, title = "Disease"),
        size = guide_legend(ncol = 2, title = "Prevotella copri")
    ) +
    theme(plot.title = element_text(hjust = 0.5))
```

# Supplemental Figure 3. Clustering scores for enterotypes in gut WGS samples.

```{r suppfig3}
pam_bray <- sapply(2:10, function(i) pam(dist_bray, k = i, cluster.only = T))
pam_js <- sapply(2:10, function(i) pam(dist_js, k = i, cluster.only = T))
pam_rjs <- sapply(2:10, function(i) pam(dist_rjs, k = i, cluster.only = T))

plot_cluster_validation = function(bray, js, rjs, legend = T, ...) {
    plot(2:10, bray, type = "b", pch = 1, xlab = "Number of Clusters", ...)
    lines(2:10, js, type = "b", pch = 2, lty = 2)
    lines(2:10, rjs, type = "b", pch = 22, lty = 3)
    if (legend) {
        legend("topright", legend = c("Bray-Curtis", "Jensen-Shannon", "Root Jensen-Shannon"), pch = c(1, 2, 22), lty = 1:3)
    }
}

ch_bray <- apply(pam_bray, 2, function(i) cluster.stats(dist_bray, i)$ch)
ch_js <- apply(pam_js, 2, function(i) cluster.stats(dist_js, i)$ch)
ch_rjs <- apply(pam_rjs, 2, function(i) cluster.stats(dist_rjs, i)$ch)

plot_cluster_validation(ch_bray, ch_js, ch_rjs, legend = T, ylab = "Calinski-Harabasz score", ylim = c(0, 300))

si_bray <- apply(pam_bray, 2, function(i) mean(silhouette(i, dist_bray)[, 3]))
si_js <- apply(pam_js, 2, function(i) mean(silhouette(i, dist_js)[, 3]))
si_rjs <- apply(pam_rjs, 2, function(i) mean(silhouette(i, dist_rjs)[, 3]))

plot_cluster_validation(si_bray, si_js, si_rjs, legend = F, ylab = "Average silhouette width", ylim = c(0, 1))
abline(0.75, 0, lty = 5, col = darkGray)
abline(0.5, 0, lty = 5, col = darkGray)
abline(0.25, 0, lty = 5, col = darkGray)
text("Strong support", x = 9, y = 0.8, col = darkGray)
text("Moderate support", x = 9, y = 0.6, col = darkGray)
text("Weak and could be artificial", x = 8.5, y = 0.4, col = darkGray)
text("No substantial structure", x = 8.5, y = 0.2, col = darkGray)

ps_bray <- prediction.strength(dist_bray, Gmin = 2, Gmax = 10, clustermethod = pamkCBI)
ps_js <- prediction.strength(dist_js, Gmin = 2, Gmax = 10, clustermethod = pamkCBI)
ps_rjs <- prediction.strength(dist_rjs, Gmin = 2, Gmax = 10, clustermethod = pamkCBI)

plot_cluster_validation(ps_bray$mean.pred[2:10], ps_js$mean.pred[2:10], ps_rjs$mean.pred[2:10], ylab = "Prediction Strength", ylim = c(0, 1.1), legend = F)
abline(0.9, 0, lty = 5, col = darkGray)
abline(0.8, 0, lty = 8, col = darkGray)
text("Strong support", x = 9, y = 1, col = darkGray)
text("Moderate support", x = 9, y = 0.85, col = darkGray)
text("Little or no support", x = 9, y = 0.6, col = darkGray)
```

# Supplemental Figure 4: Top correlations between metabolic pathways and genera.

```{r suppfig4, fig.height=8, fig.width=12}
eset.list <- curatedMetagenomicData("*pathabundance_relab.stool", dryrun = FALSE)

names(eset.list) <- gsub("\\..+", "", names(eset.list))

for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <-
        paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <-
        pData(eset.list[[i]])[,!sapply(pData(eset.list[[i]]), function(x)
            all(is.na(x)))]
    eset.list[[i]]$subjectID <-
        as.character(eset.list[[i]]$subjectID)
}

for (i in seq_along(eset.list)) {
    eset.list[[i]] <- eset.list[[i]][!grepl("\\|", rownames(eset.list[[i]])), ]
}

pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0
eset_pathway <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))

eset.list <- curatedMetagenomicData("*metaphlan_bugs_list.stool", dryrun = FALSE)

names(eset.list) <- gsub("\\..+", "", names(eset.list))

for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <-
        paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <-
        pData(eset.list[[i]])[,!sapply(pData(eset.list[[i]]), function(x)
            all(is.na(x)))]
    eset.list[[i]]$subjectID <-
        as.character(eset.list[[i]]$subjectID)
}

for (i in seq_along(eset.list)) {
    eset.list[[i]] <-
        eset.list[[i]][grep("t__", rownames(eset.list[[i]]), invert = TRUE),]
    eset.list[[i]] <-
        eset.list[[i]][grep("s__|_unclassified\t", rownames(eset.list[[i]]), perl = TRUE),]
}

pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0
eset_bugs <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))

pseq <- metaphlanToPhyloseq(tax = exprs(eset_bugs), metadat = pData(eset_bugs), split = ".")

glom_genus <- tax_glom(pseq, taxrank = "Genus")
top20_genus_otu_names = names(sort(taxa_sums(glom_genus), TRUE)[1:20])
top20_genus <- tax_table(glom_genus)[top20_genus_otu_names,"Genus"]
subset_genus <- prune_taxa(top20_genus_otu_names, glom_genus)

max_cor_pathway <- function(y, X, margin=1) {
  cors = apply(X, margin, function(x) cor.test(as.numeric(y), as.numeric(x))$estimate)
  rownames(X)[cors==max(cors, na.rm=TRUE)]
}

exprs_pwy <- exprs(eset_pathway)
exprs_pwy <- exprs_pwy[apply(exprs_pwy, 1, function(i) max(i) != 0), ]

max_pathways <- apply(otu_table(subset_genus), 1,  function(y) max_cor_pathway(y, X=exprs_pwy))

cor_matrix <- cor(t(otu_table(subset_genus)), t(exprs(eset_pathway)[max_pathways,]))
rownames(cor_matrix) = top20_genus[rownames(cor_matrix), 1]

melted_cors <- melt(cor_matrix)
melted_cors %>%
    ggplot(aes(x=Var1, y=Var2, fill=value)) +
    geom_tile() +
    scale_fill_gradient2(low = blueGreen, high = purple, mid = gray,
     midpoint = 0, space = "Lab",
     name="Pearson\nCorrelation") +
    theme(axis.text.x = element_text(angle = 45, vjust=1, hjust = 1, size=9), axis.text.y=element_text(size=9)) +
    labs(x = "Genus", y = "Pathway")
```

# Supplemental Figure 5: Alpha diversity of taxa from 11 studies of the gut microbiome.

```{r suppfig5}
alpha <- estimate_richness(pseq, measures = "Shannon")
alpha$study <- sample_data(pseq)$study

alpha %<>%
    group_by(study) %>%
    mutate(median = median(Shannon)) %>%
    arrange(desc(median)) %>%
    ungroup %>%
    mutate(study_num = as.numeric(as.factor(alpha$study)))

box_order <- factor(unique(alpha$study[order(alpha$median)]))

alpha$study <- factor(alpha$study, levels = box_order)

alpha %>%
    ggplot(aes(x = study, y = Shannon, fill = study)) +
    stat_boxplot(geom = "errorbar") +
    geom_boxplot() +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    guides(fill = guide_legend(title="Study")) +
    labs(x = "", y = "Shannon Alpha Diversity") +
    scale_fill_manual(values = c(blue, blueGreen, green, paleYellow, lightBlack, purple, red, orange, yellow, darkGray, brown))
```

# Session Info
```{r sessioninfo, cache=FALSE}
sessionInfo()
```
