---
title: "Paper Figures"
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{Paper Figures}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

```{r libraries, include=FALSE}
library(curatedMetagenomicData)
library(randomForest)
library(caret)
library(pROC)
library(ggplot2)
library(cluster)
library(fpc)
library(RColorBrewer)
library(stats)
library(reshape2)
library(utils)
```

# Coordinated Color Scheme

```{r color_scheme}
blue <- "#3366aa"
blueGreen <- "#11aa99"
green <- "#66aa55"
paleYellow <- "#cccc55"
yellow <- "#ffee33"
gray <- "#777777"
purple <- "#992288"
red <- "#ee3333"
orange <- "#ee7722"
black <- "#000000"
brown <- "#655643"
pallet <- c(purple, blue, blueGreen, green, paleYellow, yellow, orange, red, gray, brown, black)
n <- length(pallet)
image(1:n, 1, as.matrix(1:n), col = pallet, xlab = "", ylab = "", xaxt = "n", yaxt = "n", bty = "n")
```

# Supplemental Figure 1: Health status classification from species abundance.

```{r figure_1}
dataset_list <- c("KarlssonFH_2013 (T2D)", "LeChatelierE_2013 (Obesity)", "NielsenHB_2014 (IBD)", "QinJ_2012 (T2D)", "QinN_2014 (Cirrhosis)", "ZellerG_2014 (CRC)")
class_list <- c("t2d", "obesity", "ibd", "t2d", "cirrhosis", "cancer")
data_list <- c("EH277", "EH283", "EH301", "EH319", "EH325", "EH361")

eh <- ExperimentHub()

for (i in 1:length(dataset_list)) {
    taxabund <- eh[[data_list[i]]]
    
    feat <- t(exprs(taxabund))
    feat <- feat[, grep("(s__|unclassified)", colnames(feat))]
    feat <- feat[, -grep("t__", colnames(feat))]
    meta <- pData(taxabund)["disease"]
    all <- cbind(feat, meta)
    if (i == 1) {
        all <- subset(all, disease != "impaired_glucose_tolerance")
    }
    if (i == 2) {
        all <- subset(all, disease != "n")
    }
    if (i == 3) {
        all$disease[all$disease == "ibd_ulcerative_colitis"] <- "ibd"
        all$disease[all$disease == "ibd_crohn_disease"] <- "ibd"
        all$disease[all$disease == "n_relative"] <- "n"
    }
    if (i == 4) {
        all <- subset(all, disease != "na")
    }
    if (i == 6) {
        all <- subset(all, disease != "large_adenoma")
        all$disease[all$disease == "small_adenoma"] <- "n"
    }
    
    assign(paste("rf", i, sep = "_"), train(disease ~ ., data = all, preProc = c("zv", "scale", "center"), method = "rf", ntree = 500, tuneGrid = expand.grid(.mtry = c(10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 125, 150, 175, 200, 300, 400, 500)), trControl = trainControl(method = "repeatedcv", number = 10, search = "grid", summaryFunction = twoClassSummary, classProbs = TRUE, savePredictions = TRUE)))
}

for (i in 1:length(dataset_list)) {
    rf <- get(paste("rf", i, sep = "_"))
    
    if (i == 1) {
        plot.roc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry], grid = TRUE, ci = TRUE, xaxs = "i", yaxs = "i", col = pallet[i], lty = 1)
    } else {
        plot.roc(rf$pred$obs[rf$pred$mtry == rf$bestTune$mtry], rf$pred[, class_list[i]][rf$pred$mtry == rf$bestTune$mtry], grid = TRUE, ci = TRUE, xaxs = "i", yaxs = "i", col = pallet[i], lty = 1, add = TRUE)
    }
}

legend("bottomright", box.lwd = 0, box.col = "white", bg = "white", lwd = 2, legend = dataset_list, col = pallet)
```

# Supplemental Figure 2: PCoA plots colored for dataset + disease states.

```{r figure_2}
curatedMetagenomicData <- function(x = "*", dryrun = TRUE, counts = FALSE, bugs.as.phyloseq = FALSE, x.is.glob = TRUE) {
    requested.datasets <- x
    all.datasets <- ls("package:curatedMetagenomicData")
    all.datasets <- grep("marker|gene|path|metaphlan_bugs", all.datasets, val = TRUE)
    regex <- ifelse(x.is.glob, paste(glob2rx(requested.datasets), collapse = "|"), requested.datasets)
    matched.datasets <- grep(regex, all.datasets, value = TRUE)
    if (dryrun) {
        message(c(
            "Dry run: see return values for datasets that would be downloaded.",
            "\nRun with `dryrun=FALSE` to actually download these datasets."))
        return(matched.datasets)
    }
    if (!any(matched.datasets %in% all.datasets)) 
        stop("requested datasets do not match any available datasets.")
    eset.list <- lapply(seq_along(matched.datasets), function(i) {
        message(paste0("Working on ", matched.datasets[i]))
        eset <- do.call(get(matched.datasets[i]), list())
        if (counts) {
            exprs(eset) <- round(sweep(exprs(eset), 2, eset$number_reads/100, "*"))
        }
        if (bugs.as.phyloseq && grepl("metaphlan", matched.datasets[i])) {
            eset <- ExpressionSet2phyloseq(eset)
        }
        return(eset)
    })
    names(eset.list) <- matched.datasets
    return(eset.list)
}

eset.list <- curatedMetagenomicData("*metaphlan_bugs_list.stool", dryrun = FALSE)

names(eset.list) <- gsub("\\..+", "", names(eset.list))

for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <- paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <- pData(eset.list[[i]])[, !sapply(pData(eset.list[[i]]), function(x) all(is.na(x)))]
    eset.list[[i]]$subjectID <- as.character(eset.list[[i]]$subjectID)
}

for (i in seq_along(eset.list)) {
    eset.list[[i]] <- eset.list[[i]][grep("t__", rownames(eset.list[[i]]), invert = TRUE), ]
    eset.list[[i]] <- eset.list[[i]][grep("s__|_unclassified\t", rownames(eset.list[[i]]), perl = TRUE), ]
}

joinWithRnames <- function(obj, FUN = I) {
    mylist <- lapply(obj, function(x) {
        df <- data.frame(FUN(x))
        df$rnames28591436107 <- rownames(df)
        return(df)
    })
    bigdf <- Reduce(full_join, mylist)
    rownames(bigdf) <- make.names(bigdf$rnames28591436107)
    bigdf <- bigdf[, !grepl("^rnames28591436107$", colnames(bigdf))]
    return(bigdf)
}

pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))

ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0

eset <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))

metaphlanToPhyloseq <- function(tax, metadat = NULL, simplenames = TRUE, roundtointeger = FALSE, split = "|") {
    xnames <- rownames(tax)
    shortnames <- gsub(paste0(".+\\", split), "", xnames)
    if (simplenames) {
        rownames(tax) <- shortnames
    }
    if (roundtointeger) {
        tax <- round(tax * 10000)
    }
    x2 <- strsplit(xnames, split = split, fixed = TRUE)
    taxmat <- matrix(NA, ncol = max(sapply(x2, length)), nrow = length(x2))
    colnames(taxmat) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "Strain")[1:ncol(taxmat)]
    rownames(taxmat) <- rownames(tax)
    for (i in 1:nrow(taxmat)) {
        taxmat[i, 1:length(x2[[i]])] <- x2[[i]]
    }
    taxmat <- gsub("[a-z]__", "", taxmat)
    taxmat <- phyloseq::tax_table(taxmat)
    otutab <- phyloseq::otu_table(tax, taxa_are_rows = TRUE)
    if (is.null(metadat)) {
        res <- phyloseq::phyloseq(taxmat, otutab)
    } else {
        res <- phyloseq::phyloseq(taxmat, otutab, phyloseq::sample_data(metadat))
    }
    return(res)
}

pseq <- metaphlanToPhyloseq(tax = exprs(eset), metadat = pData(eset), split = ".")

samp <- data.frame(sample_data(pseq))
samp$source <- factor(samp$study == "HMP_2012", levels = c(T, F), labels = c("HMP", "Community"))
sample_data(pseq) <- samp

dist_bray <- distance(pseq, method = "bray")
dist_js <- distance(pseq, method = "jsd")
dist_rjs <- sqrt(dist_js)

ord_bray <- ordinate(pseq, method = "PCoA", distance = dist_bray)
ord_JS <- ordinate(pseq, method = "PCoA", distance = dist_js)
ord_RJS <- ordinate(pseq, method = "PCoA", distance = dist_rjs)

samp$bray_cluster_2 <- factor(pam(dist_bray, k = 2, cluster.only = T))
samp$bray_cluster_3 <- factor(pam(dist_bray, k = 3, cluster.only = T))
samp$bray_cluster_4 <- factor(pam(dist_bray, k = 4, cluster.only = T))
sample_data(pseq) <- samp

Prev <- as.numeric(otu_table(pseq)["s__Prevotella_copri", ])
samp$Prevotella_copri <- Prev
sample_data(pseq) <- samp

pc1 <- ord_bray$vectors[, 1]
pc2 <- ord_bray$vectors[, 2]

otu_tax <- attr(otu_table(pseq), "dimnames")[[1]]
otu_bacteroides <- otu_table(pseq)[grep("s__Bacteroides", otu_tax), ]
sum_bacteroides <- apply(otu_bacteroides, 2, sum)

df_ordinate <- data.frame(pc1, pc2, bact = sum_bacteroides, prev = Prev, bray2 = as.numeric(samp$bray_cluster_2) + 20)
df_bact <- df_ordinate[df_ordinate$bray2 == 21, ]
df_prev <- df_ordinate[df_ordinate$bray2 == 22, ]

samp$disease[samp$disease %in% c("obesity", "obese")] <- "obesity"
samp$disease[samp$disease %in% c("underweight", "leaness")] <- "underweight"
samp$disease_simplified[samp$disease == "cancer"] <- "cancer"
samp$disease_simplified[samp$disease %in% c("small_adenoma", "large_adenoma")] <- "adenoma"
samp$disease_simplified[samp$disease == "cirrhosis"] <- "cirrhosis"
samp$disease_simplified[samp$disease %in% c("t2d", "impaired_glucose_tolerance")] <- "t2d / impaired glucose tolerance"
samp$disease_simplified[samp$disease %in% c("ibd_crohn_disease", "ibd_ulcerative_colitis")] <- "ibd"
samp$disease_simplified[samp$disease %in% c("obesity", "overweight")] <- "obese or overweight"
df_ord_dataset_disease <- data.frame(pc1, pc2, disease_bin = factor(samp$disease == "n", levels = c(T, F), labels = c("diseased", "healthy")), disease = samp$disease_simplified, study = samp$study, prev = df_ordinate$prev)

df_ord_dataset_disease %>%
    ggplot(aes(x = pc1, y = pc2, shape = disease, color = study)) +
    geom_point() +
    labs(x = "Axis 1", y = "Axis 2", title = "PCoA by dataset and disease")
```

# Supplemental Figure 3. Clustering scores for enterotypes in gut WGS samples.

```{r figure_3}
pam_bray <- sapply(2:10, function(i) pam(dist_bray, k = i, cluster.only = T))
pam_js <- sapply(2:10, function(i) pam(dist_js, k = i, cluster.only = T))
pam_rjs <- sapply(2:10, function(i) pam(dist_rjs, k = i, cluster.only = T))

plot_cluster_validation = function(bray, js, rjs, legend = T, ...) {
    plot(2:10, bray, type = "b", pch = 1, xlab = "Number of Clusters", ...)
    lines(2:10, js, type = "b", pch = 2, lty = 2)
    lines(2:10, rjs, type = "b", pch = 22, lty = 3)
    if (legend) {
        legend("topright", legend = c("Bray-Curtis", "Jensen-Shannon", "Root Jensen-Shannon"), pch = c(1, 2, 22), lty = 1:3)
    }
}

ch_bray <- apply(pam_bray, 2, function(i) cluster.stats(dist_bray, i)$ch)
ch_js <- apply(pam_js, 2, function(i) cluster.stats(dist_js, i)$ch)
ch_rjs <- apply(pam_rjs, 2, function(i) cluster.stats(dist_rjs, i)$ch)

plot_cluster_validation(ch_bray, ch_js, ch_rjs, legend = T, ylab = "Calinski-Harabasz score", ylim = c(0, 300))

si_bray <- apply(pam_bray, 2, function(i) mean(silhouette(i, dist_bray)[, 3]))
si_js <- apply(pam_js, 2, function(i) mean(silhouette(i, dist_js)[, 3]))
si_rjs <- apply(pam_rjs, 2, function(i) mean(silhouette(i, dist_rjs)[, 3]))

plot_cluster_validation(si_bray, si_js, si_rjs, legend = F, ylab = "Average silhouette width", ylim = c(0, 1))
abline(0.75, 0, lty = 5, col = "grey70")
abline(0.5, 0, lty = 5, col = "grey70")
abline(0.25, 0, lty = 5, col = "grey70")
text("Strong support", x = 9, y = 0.8, col = "grey70")
text("Moderate support", x = 9, y = 0.6, col = "grey70")
text("Weak and could be artificial", x = 8.5, y = 0.4, col = "grey70")
text("No substantial structure", x = 8.5, y = 0.2, col = "grey70")

ps_bray <- prediction.strength(dist_bray, Gmin = 2, Gmax = 10, clustermethod = pamkCBI)
ps_js <- prediction.strength(dist_js, Gmin = 2, Gmax = 10, clustermethod = pamkCBI)
ps_rjs <- prediction.strength(dist_rjs, Gmin = 2, Gmax = 10, clustermethod = pamkCBI)

plot_cluster_validation(ps_bray$mean.pred[2:10], ps_js$mean.pred[2:10], ps_rjs$mean.pred[2:10], ylab = "Prediction Strength", ylim = c(0, 1.1), legend = F)
abline(0.9, 0, lty = 5, col = "grey70")
abline(0.8, 0, lty = 8, col = "grey70")
text("Strong support", x = 9, y = 1, col = "grey70")
text("Moderate support", x = 9, y = 0.85, col = "grey70")
text("Little or no support", x = 9, y = 0.6, col = "grey70")
```

# Supplemental Figure 4: Top correlations between metabolic pathways and genera.

```{r figure_4}
eset.list <- curatedMetagenomicData("*pathabundance_relab.stool", dryrun = FALSE)

names(eset.list) <- gsub("\\..+", "", names(eset.list))

for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <- paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <- pData(eset.list[[i]])[, !sapply(pData(eset.list[[i]]), function(x) all(is.na(x)))]
    eset.list[[i]]$subjectID <- as.character(eset.list[[i]]$subjectID)
}

for (i in seq_along(eset.list)) {
    eset.list[[i]] <- eset.list[[i]][!grepl("\\|", rownames(eset.list[[i]])), ]
}

pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0
eset_pathway <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))

eset.list <- curatedMetagenomicData("*metaphlan_bugs_list.stool", dryrun = FALSE)

names(eset.list) <- gsub("\\..+", "", names(eset.list))

for (i in 1:length(eset.list)) {
    colnames(eset.list[[i]]) <- paste(names(eset.list)[[i]], colnames(eset.list[[i]]), sep = ".")
    pData(eset.list[[i]]) <- pData(eset.list[[i]])[, !sapply(pData(eset.list[[i]]), function(x) all(is.na(x)))]
    eset.list[[i]]$subjectID <- as.character(eset.list[[i]]$subjectID)
}

for (i in seq_along(eset.list)) {
    eset.list[[i]] <- eset.list[[i]][grep("t__", rownames(eset.list[[i]]), invert = TRUE), ]
    eset.list[[i]] <- eset.list[[i]][grep("s__|_unclassified\t", rownames(eset.list[[i]]), perl = TRUE), ]
}

pdat <- joinWithRnames(eset.list, FUN = pData)
pdat$study <- sub("\\..+", "", rownames(pdat))
ab <- joinWithRnames(eset.list, FUN = exprs)
ab[is.na(ab)] <- 0
eset_bugs <- ExpressionSet(assayData = as.matrix(ab), phenoData = AnnotatedDataFrame(pdat))

pseq <- metaphlanToPhyloseq(tax = exprs(eset_bugs), metadat = pData(eset_bugs), split = ".")

glom_genus <- tax_glom(pseq, taxrank = "Genus")
top20_genus_otu_names <- names(sort(taxa_sums(glom_genus), TRUE)[1:20])
top20_genus <- tax_table(glom_genus)[top20_genus_otu_names, "Genus"]
subset_genus <- prune_taxa(top20_genus_otu_names, glom_genus)

max_cor_pathway <- function(y, X, margin = 1) {
    cors <- apply(X, margin, function(x) cor.test(as.numeric(y), as.numeric(x))$estimate)
    rownames(X)[cors == max(cors)]
}

max_pathways <- apply(otu_table(subset_genus), 1, function(y) max_cor_pathway(y, X = exprs(eset_pathway)[1:20, ]))

cor_matrix <- cor(t(otu_table(subset_genus)), t(exprs(eset_pathway)[max_pathways, ]))
rownames(cor_matrix) <- top20_genus[rownames(cor_matrix), 1]

melted_cors <- melt(cor_matrix)

melted_cors %>%
    ggplot(aes(x = Var1, y = Var2, fill = value)) +
    geom_tile() +
    scale_fill_gradient2(low = blueGreen, high = purple, mid = "gray90", midpoint = 0, space = "Lab", name = "Pearson\nCorrelation") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 9), axis.text.y = element_text(size = 9)) +
    labs(x = "Genus", y = "Pathway")
```

# Supplemental Figure 5: Alpha diversity of taxa from 11 studies of the gut microbiome.

```{r figure_5}
alpha <- estimate_richness(pseq, measures = "Shannon")
alpha$study <- sample_data(pseq)$study

alpha %<>%
    group_by(study) %>%
    mutate(median = median(Shannon)) %>%
    arrange(desc(median)) %>%
    ungroup %>%
    mutate(study_num = as.numeric(as.factor(alpha$study)))

box_order <- factor(unique(alpha$study[order(alpha$median)]))

alpha$study <- factor(alpha$study, levels = box_order)

alpha %>%
    ggplot(aes(x = study, y = Shannon, fill = study)) +
    stat_boxplot(geom = "errorbar") +
    geom_boxplot() +
    theme(axis.text.x = element_blank()) +
    ylab("Shannon Alpha Diversity") +
    xlab("")
```

# Session Info
```{r session_info}
sessionInfo()
```

