%\VignetteKeywords{Database}
%\VignetteDepends{curatedMetagenomicData}
%\VignettePackage{curatedMetagenomicData}
%\VignetteIndexEntry{curatedMetagenomicData}

\documentclass[a4paper, 10pt]{article}
\usepackage{Sweave}
\usepackage[utf8]{inputenc}
\usepackage{rotating}

<<style-Sweave, eval=TRUE, echo=FALSE, results=tex>>=
BiocStyle::latex()
@

\bioctitle[curatedMetagenomicData]{curatedMetagenomicData: Metagenomic data from different human body sites}
\author{Levi Waldron and Faizan Malik}
\begin{document}
\date{2016}

\maketitle
\tableofcontents

<<preliminaries,echo=FALSE,results=hide>>=
library(gplots)
options(width=60)
@

\section{Section 1}
Human Microbiome has played a central role in clinical research due to its impact on human health and diseases. Shotgun metagenomics sequencing technology has provided a precise way to profile microbial diversity from various human body sites across different individuals, enabling us to better understand the relationship between specific microbiota to human health and some diseases. To date, many human micro biome datasets have been generated from numerous studies as well as consortiums such as Human microbiome project \cite{Qin2010,Qin2012,Huttenhower2012, Karlsson2013,Loman2013,LeChatelier2013,Qin2014,Zeller2014,Julia2014,Nielsen2014,Alexandra2015,Rampelli2015}. Here, we present the curatedMetagenomicData Bioconductor package that contains a collection of relative abundance of taxa (ranks ranging from kingdoms to species) across over 3000 samples collected from various different human body sites as well as relative abundance of marker datasets.The package composes of two types of marker datasets:

In this vignette, we give a short tour of the package and will show how to use it efficiently.

\begin{itemize}
 \item Marker\texttt{\_}pres indicates the presence/absence of taxonomic markers, and 
 \item Marker\texttt{\_}ab contains the abundance level of markers across samples.
\end{itemize}

Marker\texttt{\_}pres datasets compose of binary values (0 and 1) that indicate whether a marker is present (1) or absent (0) in a particular sample. Marker\texttt{\_}ab datasets contains numeric values representing the abundance level of markers, where zero indicates an absence of a marker. This package provides a uniform resource for metagenomic profiles across human body sites from different studies of both published and unpublished. Having human metagenomic profiles in this format will enable an efficient identification of body sites and studies of interest for further analysis. Furthermore, an analysis of selected data set can be performed straightforwardly without the need to harmonize heterogeneous sequencing technologies, processing methods and normalizing read count data.

In this vignette, we give a short tour of the package and will show how to use it efficiently.

\section{Loading WT2D.stool.abundance data}

Loading a single dataset is very easy. First we load the package:
<<example1loadtep1>>=
library(curatedMetagenomicData)
@
To get a listing of all the datasets, use the \Rfunction{data} function:
<<example1loadtep2, eval=FALSE>>=
data(package="curatedMetagenomicData")
@
Now to load the WT2D.stool.abundance data, we use the data function again:
<<example1loadtep3, eval=TRUE>>=
data(WT2D.stool.abundance.eset) 
WT2D.stool.abundance.eset
@

The datasets are provided as Bioconductor ExpressionSet objects and we refer to the Bioconductor documentation for users unfamiliar with this data structure.

\section{Load datasets based on rules}
For a meta-analysis, we typically want to filter datasets and samples to get a set of bodysites or studies we are interested in. We provide a short R script that does the filtering and provides the data as a list of ExpressionSet objects. One can use this script within R by first sourcing a config file which specifies the filters, like the name of body site. It is also possible to filter samples by studies.

<<example2loadstep1>>=
source(system.file("extdata",
"esetselection.config",package="curatedMetagenomicData"))            
ls()
@
See what the values of these variables we have loaded are. The variable names are fairly descriptive.

\subsection{Selecting only samples from stool}
The esetselection.config file loaded above contains two objects
indicating which body site or/and study to be included. For example, if we want to select datasets that were collected only from body site ``stool'', specify the following parameters as follow:

<<example3loadSkinstep1>>=
#Specify bodysite of interest
bodysite <- "stool"
study <- ""
sapply(ls(), function(x) print(get(x)))
@ 

Now that we have defined the sample filter, we create a list of
\Rclass{ExpressionSet} objects by sourcing the \file{createEsetList.R} file:
<<example3loadSkinstep2>>=
source(system.file("extdata", "createEsetList.R", 
package ="curatedMetagenomicData"))
@

Now we have \Sexpr{length(esets)} datasets with samples that passed our filter in a list of \Rclass{ExpressionSet} objects called \texttt{esets}:
<<example3loadSkinstep3>>=
names(esets)
@

\subsection{Selecting only datasets from hmpii}
Now, if we want to select datasets that were collected only from study ``hmpii'':

<<example4loadHmpiistep1>>=
#Specify study of interest
study <- "hmpii"
bodysite <- ""
@ 

Now that we have defined the sample filter, we create a list of
\Rclass{ExpressionSet} objects by sourcing the \file{createEsetList.R} file:
<<example4loadHmpiistep2>>=
source(system.file("extdata", "createEsetList.R", 
package ="curatedMetagenomicData"))
@

Now we have \Sexpr{length(esets)} datasets with samples that passed our filter in a list of
\Rclass{ExpressionSet} objects called \texttt{esets}:
<<example4loadHmpiistep3>>=
names(esets)
@

\section{Heatmap}
Next we select all abundance datasets and plot a heatmap to investigate if the abundance of any phyla is associated with particular body sites.
<<example1prepare>>=
## load all of the abundance datasets
study <- "abundance"
bodysite <- ""
source(system.file("extdata", "createEsetList.R", 
package ="curatedMetagenomicData"))

## merge all of the eset from the resulted esets 
eset <- ExpressionSet()
for (i in 1:length(esets)){
  eset <- combine(eset, esets[[i]])
}

##Filter out features (taxa) that have low level taxonomic rank
toplt <- exprs(eset)[grep("c__", rownames(eset), invert=T),]
rel.abd.plt <- toplt

unique(eset$bodysite)
colbars <- as.integer(as.factor(eset$bodysite))
@

\begin{figure}
\centering
<<example1Heatmap, fig=TRUE>>=
res <- heatmap.2(rel.abd.plt,
          main="Abundance across samples",
          distfun=function(c) as.dist(1-cor(t(c))),
          cexRow=0.6,
          breaks=c(0, 1e-4, 0.01, 0.25, 1),
          scale="none",
          trace='none',
          key=FALSE,
          labCol=rep("", ncol(rel.abd.plt)),
          margins=c(5, 10),
          labRow=shortenCladeName(rownames(rel.abd.plt)),
          col=colorRampPalette(c("white", "black"))(4),
          ColSideColors = as.character(colbars)
)
@
\caption{Heatmap showing abundance of phyla across samples. This heatmap only lists features at the kingdom and phylum level to get an idea about which phyla are enriched across different body sites. The columns in the heat map represent different body sites. It shows that Firmicutes, Actinobacteria, and Proteobacteria are the most prominent phylum in human metagenomics. Black color = 100 \% abundance, White color = 0 \% abundance}
\label{fig:example1:heatmap}
\end{figure}

\clearpage
\bibliography{curatedMetagenomicData_vignette}

\SweaveOpts{concordance=TRUE}

\end{document}

