%\VignetteKeywords{Database}
%\VignetteDepends{curatedMetagenomicData}
%\VignettePackage{curatedMetagenomicData}
%\VignetteIndexEntry{curatedMetagenomicData}

\documentclass[a4paper, 10pt]{article}
\usepackage{Sweave}
\usepackage[utf8]{inputenc}
\usepackage{rotating}

<<style-Sweave, eval=TRUE, echo=FALSE, results=tex>>=
BiocStyle::latex()
@

\bioctitle[curatedMetagenomicData]{curatedMetagenomicData: Metagenomic data from different human body sites}
\author{Levi Waldron and Faizan Malik}
\begin{document}
\date{2016}

\maketitle
\tableofcontents

<<preliminaries,echo=FALSE,results=hide>>=
library(gplots)
options(width=60)
@

\section{Introduction} 

Human Microbiome has played a central role in clinical research due to its
impact on human health and diseases. Shotgun metagenomics sequencing technology
has provided a precise way to profile microbial diversity from various human
body sites across different individuals, enabling us to better understand the
relationship between specific microbiota to human health and some diseases. To
date, many human micro biome datasets have been generated from numerous studies
as well as consortiums such as Human microbiome project
\cite{Qin2010,Qin2012,Huttenhower2012,
Karlsson2013,Loman2013,LeChatelier2013,Qin2014,Zeller2014,Julia2014,Nielsen2014,Alexandra2015,Rampelli2015}.
Here, we present the curatedMetagenomicData Bioconductor package that contains
a collection of relative abundance of taxa (ranks ranging from kingdoms to
species) across over 3000 samples collected from various different human body
sites as well as relative abundance of marker datasets.The package composes of
two types of marker datasets:

\begin{itemize}
  \item Marker\texttt{\_}pres indicates the presence/absence of taxonomic 
    markers, and 
  \item Marker\texttt{\_}ab contains the abundance level of markers across 
    samples.
\end{itemize}

Marker\texttt{\_}pres datasets compose of binary values (0 and 1) that indicate
whether a marker is present (1) or absent (0) in a particular sample.
Marker\texttt{\_}ab datasets contains numeric values representing the abundance
level of markers, where zero indicates an absence of a marker. This package
provides a uniform resource for metagenomic profiles across human body sites
from different studies of both published and unpublished. Having human
metagenomic profiles in this format will enable an efficient identification of
body sites and studies of interest for further analysis. Furthermore, an
analysis of selected data set can be performed straightforwardly without the
need to harmonize heterogeneous sequencing technologies, processing methods and
normalizing read count data.

In this vignette, we give a short tour of the package and will show how to use
it efficiently.

\section{Data discovery}

The data are hosted in the Bioconductor ExperimentHub. We can use functions from
the ExperimentHub package to list and load resources specific to 
\Rcode{curatedMetagenomicData}.

First load the package:
<<example1loadtep1>>=
library(curatedMetagenomicData)
@

\Rfunction{listResources} returns the names of all resources in 
the package:
<<example1loadtep2>>=
allData <- listResources(ExperimentHub(), "curatedMetagenomicData")
length(allData)
head(allData)
@

The query can be filtered by terms specific to the resource such as
study, bodysite or data type. 
<<example1loadtep2>>=
listResources(ExperimentHub(), "curatedMetagenomicData", filterBy="throat")
@

The name of each resource can be invoked as a function. The \Rcode{metadata}
argument controls whether metadata are returned or if the data are
loaded. 

\Rcode{metadata=TRUE} returns just the metadata:
<<example1loadtep3, eval=TRUE>>=
WT2D.stool.abundance.eset(metadata=TRUE)
@

\Rcode{metadata=FALSE} loads the dataset:
<<example1loadtep4, eval=TRUE>>=
WT2D.stool.abundance.eset(metadata=FALSE)
@

Data can also be loaded with the \Rcode{loadResources} function. The
\Rcode{filterBy} argument can be the name of a resource or any term in the
resource metadata.
<<example1loadtep5, eval=TRUE>>=
loadResources(ExperimentHub(), "curatedMetagenomicData", filterBy="WT2D.stool.abundance")
@

The datasets are provided as Bioconductor \Rclass{ExpressionSet} objects and we
refer to the Bioconductor documentation for users unfamiliar with this data
structure.

\section{Filtering datasets for analysis} 

For a meta-analysis, we typically want to filter datasets and samples to get a
set of bodysites or studies we are interested in. \Rcode{loadResources} 
filters resources based on the values in \Rcode{filterBy} then loads the 
datasets into a list.

Any term in the resource metadata can be used as a filter term:
<<>>=
t2dmeta_long.stool.pathcoverage.eset(metadata=TRUE)
@

Body type, study and data type terms can be seen in the resource titles:
<<>>=
head(listResources(ExperimentHub(), "curatedMetagenomicData"))
@

List all datasets that were collected only from body site ``saliva":
<<>>=
filtered <- listResources(ExperimentHub(), "curatedMetagenomicData", filterBy="saliva")
filtered
@ 

Load the data as a list of \Rclass{ExpressionSet} objects:
<<>>=
esets <- loadResources(ExperimentHub(), "curatedMetagenomicData", filterBy=filtered)
sapply(esets, class)
@

We could have accomplished the same thing by passing ``saliva" as the
\Rcode{filteredBy} argument:
<<eval=FALSE>>=
loadResources(ExperimentHub(), "curatedMetagenomicData", filterBy="saliva")
@

Search terms can be combined for more detailed filtering. Here we list all
datasets from the ``hmp" study:

<<example4loadHmpiistep1>>=
length(listResources(ExperimentHub(), "curatedMetagenomicData", "hmp"))
@ 

The results can be narrowed down by filtering by the body site 
``hard{\_}palate":
<<example4loadHmpiistep2>>=
listResources(ExperimentHub(), "curatedMetagenomicData", c("hmp", "hard_palate"))
@

Load the resources into a list:
<<example4loadHmpiistep3>>=
hmp <- loadResources(ExperimentHub(), "curatedMetagenomicData", c("hmp", "hard_palate"))
names(hmp)
@

We have \Sexpr{length(hmp)} datasets with samples that passed our filter
in a list of \Rclass{ExpressionSet} objects called \texttt{hmp}:


\section{Heatmap}

Next we select all abundance datasets and plot a heatmap to investigate if the
abundance of any phyla is associated with particular body sites.
<<example1prepare1>>=
## load all of the abundance datasets
esets <- loadResources(ExperimentHub(), "curatedMetagenomicData", "\\.abundance")

## merge all ExpressionSet objects into one
eset <- ExpressionSet()
for (i in 1:length(esets)){
  eset <- combine(eset, esets[[i]])
}
@

Filter out features (taxa) that have low level taxonomic rank:
<<example1prepare2>>=
toplt <- exprs(eset)[grep("c__", rownames(eset), invert=TRUE),]
rel.abd.plt <- toplt

unique(eset$bodysite)
colbars <- as.integer(as.factor(eset$bodysite))
@

\begin{figure}
\centering
<<example1Heatmap, fig=TRUE>>=
dist.na.removed <- function(c) {
  dist <- as.dist(1-cor(t(c)))
  dist[is.na(dist)] <- 0  
  dist
}

res <- heatmap.2(rel.abd.plt,
          main="Abundance across samples",
          distfun=function(c) dist.na.removed(c),
          cexRow=0.6,
          breaks=c(0, 1e-4, 0.01, 0.25, 1),
          scale="none",
          trace='none',
          key=FALSE,
          labCol=rep("", ncol(rel.abd.plt)),
          margins=c(5, 10),
          labRow=curatedMetagenomicData:::.shortenCladeName(rownames(rel.abd.plt)),
          col=colorRampPalette(c("white", "black"))(4),
          ColSideColors = as.character(colbars)
)
@
\caption{Heatmap showing abundance of phyla across samples. This heatmap only
lists features at the kingdom and phylum level to get an idea about which phyla
are enriched across different body sites. The columns in the heat map represent
different body sites. It shows that Firmicutes, Actinobacteria, and
Proteobacteria are the most prominent phylum in human metagenomics. Black color
= 100 \% abundance, White color = 0 \% abundance}
\label{fig:example1:heatmap}
\end{figure}

\clearpage
\bibliography{curatedMetagenomicData_vignette}

\SweaveOpts{concordance=TRUE}

\end{document}
