---
title: "Analyzing curated metagenomic data through ExperimentHub"
author: "Faizan Malik, Levi Waldron"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    number_sections: yes
    toc: yes
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{curatedOvarianData through ExperimentHub}
---

A built [html][] version of this vignette is available.

```{r, echo=FALSE, warning=FALSE}
suppressPackageStartupMessages(library(ExperimentHub))
```

# What `curatedMetagenomicData` provides

`curatedMetagenomicData` provides 6 types of data for each dataset:

1. species-level taxonomic profiles,
2. presence of unique, clade-specific markers,
3. abundance of unique, clade-specific markers,
4. abundance of gene families,
5. metabolic pathway coverage, and 
6. metabolic pathway abundance. 

Types 1-3 are generated by [MetaPhlAn2](http://huttenhower.sph.harvard.edu/metaphlan2); 4-6 are generated by [HUMAnN2](http://huttenhower.sph.harvard.edu/humann2).

Currently, `curatedMetagenomicData` provides:

* 2,875 samples from 13 datasets, primarily of the human gut but including body sites profiled in the Human Microbiome Project
* processed data from whole-metagenome shotgun metagenomics, with manually-curated metadata, as integrated and documented Bioconductor ExpressionSet objects
* ~200 fields of specimen metadata from original papers, supplementary files, and websites, with manual curation to standardize annotations
* processing of data through the new MetaPhlAn2 pipeline  (http://huttenhower.sph.harvard.edu/metaphlan2) for taxonomic abundance, and HUMAnN2 pipeline (Version 2 manuscript in preparation, http://huttenhower.sph.harvard.edu/humann2) for metabolic analysis. 
* This effort required analyzing ~23T of raw sequencing data, generating ~52T of intermediate files

These datasets are documented in the [reference manual](https://github.com/vobencha/curatedMetagenomicData/files/341411/curatedMetagenomicData.pdf).

# Setup

Load the ExperimentHub library, 

```{r, message=FALSE, cache=FALSE}
library(ExperimentHub)
```

and create the hub:

```{r}
eh = ExperimentHub()
```

The following queries ExperimentHub for all records matching "curatedMetagenomicData":

```{r}
myquery = query(eh, "curatedMetagenomicData")
```

This is an abbreviated list of what was found:

```{r}
myquery
```

Note that the first column, "EH2", "EH3", for example, are unique IDs for each dataset. The low numbers here are because `curatedMetagenomicData` is the first major dataset to be distributed through `ExperimentHub`. For a full list, `mcols(myquery)` produces a `data.frame`. The following is unevaluated in this script, but can be used to display an interactive spreadsheet of the results:

```{r, eval=FALSE}
View(mcols(myquery))
```

Or to write the search results to a csv file:

```{r}
write.csv(mcols(myquery), file="curatedMetagenomicData_allrecords.csv")
```

# Choose a dataset for analysis

Here we choose "EH2" from the View(mcols(myquery)) above. This loads the "Candela_Africa" stool dataset. Note that this operation will occur much more quickly subsequent times that it is performed, due to local caching.

Please ignore the warning that this generates, it is harmless and will not appear in the release version.

```{r, warning=FALSE}
taxabund = eh[["EH2"]]
taxabund
```

This `ExpressionSet` object contains information about the experiment:

```{r}
experimentData(taxabund)@name
experimentData(taxabund)@title
experimentData(taxabund)@abstract
experimentData(taxabund)@other
```

It also contains information about the study subjects. For example, here are the first 11 columns of data (all that are available for this dataset), for the first 5 subjects:

```{r}
pData(taxabund)[1:5, 1:11]
```

Again, this could be explored more fully by (unevaluated):

```{r, eval=FALSE}
View(pData(taxabund))
```

And of course, it contains the table of taxonomic abundances. Here are the relative abundances for the first two Operational Taxonomic Units (OTUs), for the first five subjects:

```{r}
exprs(taxabund)[1:2, 1:5]
```

We refer users to the [MetaPhlAn2 documentation](http://huttenhower.sph.harvard.edu/metaphlan2) for description of what these values represent.

# Choose metabolic datasets for analysis

Recall that we looked up dataset codes such as "EH6" by doing `View(mcols(myquery))`.

Let's look, for example, at pathway coverage for the Candela Africa dataset:

```{r, message=FALSE, warning=FALSE}
pathcoverage = myquery[["EH7"]]
```

Coverage of the first two pathways, for the first two subjects, is:

```{r}
exprs(pathcoverage)[1:2, 1:2]
```

We refer users to the [HUMAnN2 documentation](http://huttenhower.sph.harvard.edu/humann2) for description of what these values represent.

# Taxonomy-aware analysis with the `phyloseq` Bioconductor library

`curatedMetagenomicData` provides 6 types of data; of these, only the taxonomic profiles make sense to analyse as phylogenetic data. Here we use a convenience script from the Waldron lab to create the `phyloseq` object

```{r, cache=FALSE, message=FALSE}
library(phyloseq)
source("https://raw.githubusercontent.com/waldronlab/presentations/master/Waldron_2016-06-07_EPIC/metaphlanToPhyloseq.R")
```

And create the `phyloseq` object from the above-created object "taxabund".

```{r}
Candela = metaphlanToPhyloseq(tax = exprs(taxabund), 
                              metadat = pData(taxabund),
                              simplenames = TRUE, 
                              roundtointeger = TRUE)
```

Now we can do all sorts of taxonomy-aware analyses of this dataset...


```{r, eval=TRUE}
summary(otu_table(Candela))
summary(sample_data(Candela))
```

## Initial exploration

`phyloseq` help vignettes [here](https://bioconductor.org/packages/release/bioc/html/phyloseq.html).

```{r}
Candela
Candela = prune_taxa(taxa_sums(Candela) > 0, Candela)
Candela
```

## Subsetting and pruning
```{r}
rank_names(Candela)
subset_taxa(Candela, !is.na(Strain))
(Candela.sp_strain = subset_taxa(Candela, !is.na(Species)))

taxonomy.level = apply(tax_table(Candela), 1, function(x) sum(!is.na(x)))
Candela.phy = prune_taxa(taxonomy.level==2, Candela)
taxa_names(Candela.phy)

```

## Advanced pruning

Keep taxa only if they are in the most abundant 10% of taxa in at least two samples:
```{r}
f1<- filterfun_sample(topp(0.1))
pru <- genefilter_sample(Candela, f1, A=2)
summary(pru)
subset_taxa(Candela, pru)
```

More help [here](https://bioconductor.org/packages/release/bioc/vignettes/phyloseq/inst/doc/phyloseq-basics.html#trimming-subsetting-filtering-phyloseq-data).

## Heatmap

```{r}
plot_heatmap(Candela.sp_strain, method="PCoA", distance="bray")
```

## Barplot

```{r}
par(mar = c(20, 4, 0, 0) + 0.1) # make more room on bottom margin
barplot(sort(taxa_sums(Candela.sp_strain), TRUE)[1:30]/nsamples(Candela.sp_strain), las=2)
```

## Alpha diversity estimates

* Look at `?phyloseq::estimate_richness`
* Supported measures of alpha diversity are:
    - "Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher"
    - more information from `vegan` package

Note, you can ignore warning about singletons:
```{r, warning=FALSE}
alpha_meas = c("Shannon", "Simpson", "InvSimpson")
(p <- plot_richness(Candela, "gender", "camp", measures=alpha_meas))
```

## Comparison of alpha diversity estimates

```{r}
alphas = estimate_richness(Candela, measures=alpha_meas)
pairs(alphas)
```

## Beta diversity / dissimilarity 

E.g. Bray-Curtis dissimilarity between all pairs of samples:

```{r}
plot(hclust(phyloseq::distance(Candela, method="bray")), 
     main="Bray-Curtis Dissimilarity", xlab="", sub = "")
```

* Dozens of distance measures are available
    - see `?phyloseq::distance` and `?phyloseq::distanceMethodList`

## Ordination

```{r}
ord = ordinate(Candela, method="PCoA", distance="bray")
plot_ordination(Candela, ord, color="camp", shape="camp") + 
  ggplot2::ggtitle("Bray-Curtis Principal Coordinates Analysis")
```

* Available methods are "DCA", "CCA", "RDA", "CAP", "DPCoA", "NMDS", "MDS", "PCoA"
* PCoA approximately preserves distances in a two-dimensional projection


Screeplot:

```{r}
plot_scree(ord) + ggplot2::ggtitle("Screeplot")
```

# sessionInfo()
```{r}
sessionInfo()
```

[html]: http://rpubs.com/lwaldron/curatedmetagenomicdata
